---
title: "ProjectBroadAnnots.Rmd"
author: "Jenna LaBelle"
date: "10/07/22"
output: html_document
---

## Expression patterns of various genes of interest acros PD1 treated samples/untreated GBM

```{r}
library(Seurat) 
library(ggplot2)
library(ggpubr)

base_dir = "/Users/jlabelle/"
script_dir = "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Scripts/"
working_dir<- paste0(base_dir, 
                     "Dropbox (Partners HealthCare)/FilbinLab/",
                     "data_analysis/Projects/git_repos/Pedi_Immune/PD1_Treat/comparisonToUntreated/")

analysis_dir<- paste0(working_dir, "/analysis/goi_expression/")
figure_dir<- paste0(working_dir, "/figures/goi_expression/")
data_dir<- paste0(working_dir, "analysis/")

if(!dir.exists(figure_dir)){dir.create(figure_dir, recursive = TRUE)}
if(!dir.exists(analysis_dir)){dir.create(analysis_dir, recursive = TRUE)}

preprocessing = paste0(base_dir, script_dir, "single_cell_preprocessing_helper_functions.R")
nmf_helper = paste0(base_dir, script_dir, "NMF_helper_function.R")
source(preprocessing)
source(nmf_helper)
source("../HelperFunctions.R")

```


## Load seurat objects: pd-1 (processed in 01b_Preprocess_QC_ToSeurat.Rmd) and full immune cohort (both myeloid and t cells)

```{r}
## Set whether to use integrated or unintegrated objects
integration<- "Harmony" ## NoInt or Harmony- as of 11/1/22, not using seurat integration (SeuratInt) downstream
figure_dir_use<- paste0(figure_dir, integration, "/")
if(!dir.exists(figure_dir_use)){dir.create(figure_dir_use)}

full_cohort_dir<- "../../ImmuneCellAnalysis/"


## Load list of merged ss2/10x seurat objects (6 total- cd4/cd8/myeloid for tenx and ss2)
ss2_seurat<- readRDS("analysis/project_programs/ss2_seurat_list_cd4.cd8.myeloid.Rds")
tenx_seurat<- readRDS("analysis/project_programs/tenx_seurat_list_cd4.cd8.myeloid.Rds")

## Colors 
tcell_colors<- readRDS(paste0(full_cohort_dir, "/plot_colors/Tcell.programs_rank8sharedByunion_reAnnot.Rds"))
myeloid_colors<- readRDS(paste0(full_cohort_dir, "/plot_colors/Myeloid_program_colors.Rds"))
myeloid_colors<- myeloid_colors$rank6
immune_colors<- c(tcell_colors, myeloid_colors)
colors_treatment<- c(PD1="orange", Untreated="purple")

  
## Nature 2020 genesets
nature_2020_genes<- read.csv(paste0(base_dir, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Shared/Marker_genes/Immune/Published/",
                                    "Nature2020_genesets.csv"))
colnames(nature_2020_genes)<- c("cytotoxic", "naive", "predysfunctional", "dysfunctional")
nature_2020_genes<- as.list(nature_2020_genes)
nature_2020_genes<- lapply(nature_2020_genes, function(x){x[x!=""]})

```

## Settings:
## set GOI to plot, sample order (based on treatment, sequencing), whether to plot programs assigned by score or projection
```{r}
GOI<- c("KLRB1", "SELPLG", "PTGER2", "PDCD1", "TIGIT", "CTLA4", "ITGAE", "GZMK", "GNLY", "NKG7", "GZMB", "GZMA", "GADD45B")

sample_order<-c("BT1478", "BT1857","BT1745", "E167", "BT1910", "BT1935", "GEX")

## Can assign programs by score or by program- assign which to use here. Will set metadata column to use downstream + new figure dir
annot_by<- "score" ## "score" or "program
if(annot_by)
```

## Merge by sequencing type
```{r}
all_seurat<- lapply(gsub("tenx_", "", names(tenx_seurat)), function(x){
  merge(tenx_seurat[[paste0("tenx_", x)]], ss2_seurat[[paste0("ss_", x)]])
})
names(all_seurat)<-gsub("tenx_", "", names(tenx_seurat))

```

## Pseudobulked heatmap of expression of GOI
```{r}
all_plots<- lapply(gsub("tenx_", "", names(tenx_seurat)), function(x){
  tmp<- all_seurat[[x]]
  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$sample, "_", gsub("\\.", "\n", gsub("_", "\n", tmp$Treatment)))
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  pb_center<- NormCenter(pb)$center_data
  
  ## Plot
  myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE,
            orderSubtypes =sample_order,
            orderFactors = c("Untreated", "PD1")) +
    scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=3)
ggsave(paste0(figure_dir, "Heatmap_GOI_All.png"), width=25, height=7)
```


## Pseudobulked heatmap of expression of GOI, by programs
```{r}
all_plots<- lapply(c("cd4", "cd8"), function(x){
  tmp<- all_seurat[[x]]
  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$sample, "_", gsub("\\.", "\n", gsub("_", "\n", tmp$Program_projected)))
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  pb_center<- NormCenter(pb)$center_data
  
  ## Plot
  myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE,
            orderSubtypes  = sample_order) +
    scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=2)
ggsave(paste0(figure_dir, "Heatmap_GOI_byProgram_All.png"), width=28, height=7)
```

## Based on heatmap- boxplot of mean gene expression for GOI within program of interest
```{r}
tmp<- all_seurat$cd4
tmp$Program_projected<- gsub("\\.", "\n",tmp$Program_projected)

## Remove effector memory- no untreated GBM effector memory (+ stress for cd4)
tmp<- subset(tmp, Program_projected != "Effector\nMemory")
tmp<- subset(tmp, Program_projected != "Stress")

## Boxplot
goi_plot<- c("PDCD1", "GNLY", "TIGIT", "CTLA4", "GZMB", "SELPLG", "PTGER2", "KLRB1","GADD45B")
all_plots<-lapply(goi_plot, function(gene){
  tmp$GOI<- as.data.frame(tmp@assays$RNA@data[gene,])
  mean_goi<- tmp@meta.data %>% group_by(sample, Program_projected) %>% summarise(mean_expr=mean(GOI)) %>% as.data.frame()
  mean_goi$Treatment<- plyr::mapvalues(mean_goi$sample, tmp$sample, tmp$Treatment, warn_missing = FALSE)
  mean_goi$Treatment<- factor(mean_goi$Treatment, levels=c("Untreated", "PD1"))
  
  ggplot(mean_goi, aes(x=Treatment, y=mean_expr, fill=Treatment))+
    geom_boxplot()+
    geom_point()+
    ylim(0,max(mean_goi$mean_expr))+
    facet_grid(cols=vars(Program_projected))+
    stat_compare_means(method="t.test")+
    theme_bw()+
    scale_fill_manual(values=c(Untreated="purple", PD1="orange"))+
    theme(axis.text.x = element_text(angle=45, hjust=1),
          axis.text = element_text(face="bold", color="black"),
          axis.title = element_text(face="bold", color="black"),
          strip.text = element_text(face="bold", color="black", size=10),
          legend.position = "none")+
    xlab("")+
    ylab(paste0("Mean log2 TPM: ", gene))
})
cowplot::plot_grid(plotlist=all_plots, ncol=1)
ggsave(paste0(figure_dir, "Boxplot_AllGOI_MeanExprBySample_PD1vUntreated_CD4.png"), width=10, height=4*length(goi_plot))

## VlnPlot
VlnPlot(tmp, group.by = "Program_projected",stack=TRUE , flip = TRUE,
        features=c("PDCD1", "TIGIT", "CTLA4", "ITGAE", "GZMK", "GNLY", "NKG7", "GZMA", "SELPLG", "PTGER2","KLRB1", "GADD45B"), 
        combine=TRUE, fill.by="ident", split.by = "Treatment", split.plot = TRUE)+
  scale_fill_manual(values= c(Untreated="purple", PD1="orange"))
ggsave(paste0(figure_dir, "Vln_GOI_byTreatment_CD4.png"), width=7, height=8)
```

## Score for cytoxic/dysfunctional genesets, plot by treatment
```{r}
## Add nature 2020 scores (dysfunctional, cyto, dys, memory)
all_seurat<- lapply(all_seurat, function(x){
  x<-AddModuleScore(x, nature_2020_genes, name=paste0(names(nature_2020_genes), "_score"))
  colnames(x@meta.data)<-ifelse(grepl(paste(paste0(names(nature_2020_genes), "_score"), collapse = "|"), colnames(x@meta.data)),
                                gsub("[0-9]", "", colnames(x@meta.data)),
                                colnames(x@meta.data))
  return(x)
})

## Plot
all_plots<- lapply(names(all_seurat), function(x){
  seurat<- all_seurat[[x]]
  seurat$sample<- factor(seurat$sample, sample_order[sample_order %in% unique(seurat$sample)])
  DotPlot(seurat, group.by = "sample", features=paste0(names(nature_2020_genes), "_score"))+
    scale_color_gradient2(low="blue", mid="grey", high="red")+
    coord_flip()+
    theme(axis.text.x = element_text(angle=45, hjust=1))+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=1)
ggsave(paste0(figure_dir, "DotPlot_TcellStateScores_CD4.CD8.png"), width=13, height=8)
```

