---
title: "velocyto"
output: html_document
---

```{r read-data}
knitr::opts_chunk$set(echo = TRUE)

library(loomR)
library(velocyto.R)
library(magrittr)
library(destiny)

base_dir = "/Users/jlabelle/"
script_dir = "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Scripts/"
working_dir<- paste0(base_dir, 
                     "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Projects/git_repos/Pedi_Immune/",
                     "TumorCellAnnotation/BySubtype/Hemispheric/BCOR/")
file_dir<- paste0(working_dir, "analysis/velocyto/loomFiles/")

preprocessing = paste0(base_dir, script_dir, "single_cell_preprocessing_helper_functions.R")
nmf_helper = paste0(base_dir, script_dir, "NMF_helper_function.R")
source(preprocessing)
source(nmf_helper)

marker_dir = "/Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Marker_genes/"

#color_helper = paste0(base_dir, script_dir, "colors.R")
#source(color_helper)

liger_fig_dir<- paste0(working_dir, "figures/Liger/")
liger_analysis_folder<- paste0(working_dir, "analysis/liger/")
velocyto_analysis_folder = "analysis/velocyto/"
velocyto_fig_folder = "figures/velocyto/"
for (folder in c(velocyto_analysis_folder, velocyto_fig_folder)){
  if (!dir.exists(folder)){
    dir.create(folder)
  }
}

```

# Data Loading

## Load loom data
```{r load_loom_data}
loom_files = list.files(path = file_dir, pattern="*.loom")

loom_files_list_all = lapply(loom_files, function(x) read.loom.matrices(paste0(file_dir, x)))
names(loom_files_list_all)<- gsub("onefilepercell_", "", gsub("-.*", "", loom_files))
```

## Simplify cell names 
```{r }
loom_files_list_all = lapply(loom_files_list_all, 
                         function(ldat) ldat <- lapply(ldat,function(x) {
                           colnames(x) <-  gsub(".genome.bam","",gsub(".*:","",colnames(x)))
                           x}
                           )
                         )
```


## Optional: subset to 1 sample 
```{r}
loom_files_list_BT1478<-list(loom_files_list_all$BT1478)
loom_files_list_BT1516<- list(loom_files_list_all$BT1516)
loom_files_list_MUV6566<- list(loom_files_list_all$MUV65, loom_files_list_all$MUV66)
loom_files_list_NoBT1478<- list(loom_files_list_all$BT1516,
                                loom_files_list_all$MUV65, 
                                loom_files_list_all$MUV66)

loom_files_list<- loom_files_list_MUV6566
sample<- c( "MUV65", "MUV66")
```

## Concatenate each matrix together 
```{r}
# exonic read (spliced) expression matrix
emat = NULL
# intronic read (unspliced) expression matrix
nmat = NULL
# spanning read (intron+exon) expression matrix
smat = NULL
for (i in 1:length(loom_files_list)){
  emat = cbind(emat, loom_files_list[[i]]$spliced)
  nmat = cbind(nmat, loom_files_list[[i]]$unspliced)
  smat = cbind(smat, loom_files_list[[i]]$spanning)
}
```

## Add preceding 0 to cell names
```{r}
colnames(emat) = sapply(colnames(emat), function(x) ifelse(grepl("MUV0.", x), x, gsub("MUV", "MUV0", x)))
colnames(nmat) = sapply(colnames(nmat), function(x) ifelse(grepl("MUV0.", x), x, gsub("MUV", "MUV0", x)))
colnames(smat) = sapply(colnames(smat), function(x) ifelse(grepl("MUV0.", x), x, gsub("MUV", "MUV0", x)))
```

## Replace "-" with "." in cell names 
```{r}
colnames(emat) = gsub("-", ".", colnames(emat), fixed=T)
colnames(nmat) = gsub("-", ".", colnames(nmat), fixed=T)
colnames(smat) = gsub("-", ".", colnames(smat), fixed=T)
```


## Generate cell colors and embedding from seurat 
```{r make_embedding}
load(paste0("analysis/seurat/seurat_obj_", "MUV65_MUV66", ".Robj"))
#cells_use<- seurat_obj@meta.data$sample %in% sample

## Optional: further subsetting of cells
  ## Current NMF
  cells_use<- cells_use & seurat_obj@meta.data$signature_1 !="NMF1"

  ## Liger NMF
  nmf_score<-readRDS(paste0(liger_analysis_folder, "nmf_score_AllFactors", ".Rds"))
  cells_use<- cells_use & (nmf_score$signature_1=="factor13" | 
                             nmf_score$signature_1 == "factor3" |
                             nmf_score$signature_1== "factor8")
  cells_use_nmf<- (nmf_score$signature_1=="factor9" | 
                             nmf_score$signature_1 == "factor10" |
                             nmf_score$signature_1== "factor14") & rownames(nmf_score) %in% colnames(seurat_obj)
  cells_use_nmf<-  rownames(nmf_score) %in% colnames(seurat_obj)
  


## Cell names
#cell_names = colnames(seurat_obj)[cells_use]
cell_names = colnames(seurat_obj)[colnames(seurat_obj) %in%  rownames(nmf_score)[cells_use_nmf]]
seurat_obj<- subset(seurat_obj, cells=cell_names)

## UMAP embedding 
emb = seurat_obj@reductions$umap@cell.embeddings

## Harmony embedding
emb_harmony = seurat_obj@reductions$harmony@cell.embeddings[,1:2]

## NMF scores
emb_NMF = as.matrix(seurat_obj@meta.data[,c("NMF1", "NMF2", "NMF3", "NMF4", "NMF5")])[cells_use,]
```




## Subset emat/nmat/smat with cells in embedding 
```{r}
## Check if any cells in the embedding are not present in the count matrix
any(!(cell_names %in% colnames(emat)))

emat = emat[,cell_names]
nmat = nmat[,cell_names]
smat = smat[,cell_names]
```
## Colors for velocyto plot 
```{r}
## Read in nmf scores, assign color to each. Will not change even when some programs removed

  ## current nmf
  nmf_score = readRDS("analysis/nmf/Merge/nmf_scoreMerge1.rds")
  gg_colors_all = c("deepskyblue", "green", "blue", "magenta", "red")
  names(gg_colors_all)<- sort(unique(nmf_score$signature_1))
  seurat_meta_toUse<- seurat_obj@meta.data[cells_use,]
  
  ## liger inmf
  nmf_score<-readRDS(paste0(liger_analysis_folder, "nmf_score_AllFactors", ".Rds"))
  gg_colors_all = c("deepskyblue", "green", "blue", "magenta", "red", "grey", "darkgreen", "darkorchid4")
  names(gg_colors_all)<- sort(unique(nmf_score$signature_1))
  
  #seurat_obj_liger<- subset(seurat_obj, cells=colnames(seurat_obj)[cells_use]) 
  seurat_obj_liger<- AddMetaData(seurat_obj, nmf_score[cells_use_nmf,]) ## adding sig 1 from LIGER to seurat
  seurat_meta_toUse<- seurat_obj_liger@meta.data

## Subset scores to just scores for cells of interest
nmf_score = nmf_score[cell_names,]

## Make colors for metagene programs
cluster_numbers = sort(unique(nmf_score$signature_1))
gg_colors<- gg_colors_all[names(gg_colors_all) %in% cluster_numbers]


cell.colors = as.character(plyr::mapvalues(x = nmf_score$signature_1, 
                                           from = cluster_numbers, 
                                           to = gg_colors))
names(cell.colors) = cell_names

colors = list()
for (i in 1:length(cluster_numbers)){
  colors[[cluster_numbers[i]]] = gg_colors[i] 
}
#save(colors, file=paste0(velocyto_analysis_folder, "colors.Robj"))

## Make colors for samples
sample_colors = gg_color_hue(length(unique(seurat_meta_toUse$sample)))
sample_numbers = sort(unique(seurat_meta_toUse$sample[cells_use]))
cell.colors.samples = as.character(plyr::mapvalues(x = seurat_meta_toUse$sample[cells_use], 
                                                   from = sample_numbers, to = sample_colors))
names(cell.colors.samples) = cell_names

## Make colors for subtypes 
subtype_colors = gg_color_hue(length(unique(seurat_meta_toUse$subtype)))
subtype_numbers = sort(unique(seurat_meta_toUse$subtype))
cell.colors.subtypes = as.character(plyr::mapvalues(x = seurat_meta_toUse$subtype, from = subtype_numbers, to = subtype_colors))
names(cell.colors.subtypes) = cell_names

## Make colors for cell-cycle
cc_binary = seurat_meta_toUse$signature_1[cells_use]=="NMF1"
cc_colors = c("lightgrey", "red3")
cc_numbers = sort(unique(cc_binary))
cell.colors.cc = as.character(plyr::mapvalues(x = cc_binary, from =cc_numbers, to = cc_colors))
names(cell.colors.cc) = cell_names

## Reclassify cycling as second top program, then make colors for new vecotr
cyclingFactor<- c("factor6", "factor7")
cycling_cells<- seurat_meta_toUse[seurat_meta_toUse$signature_1 %in% cyclingFactor,]

  ## reclassify to sig2, or sig3 if sig2 is also a cycling factor
  cycling_reclassify<- c()
  for (i in 1:nrow(cycling_cells)){
    sig2<- cycling_cells$signature_2[i]
    if (sig2 %in% cyclingFactor){
      sig3<- cycling_cells$signature_3[i]
      cycling_reclassify<- c(cycling_reclassify, sig3)
    } else {cycling_reclassify<- c(cycling_reclassify, sig2)}
  }
  names(cycling_reclassify)<- rownames(cycling_cells)

  ## Make list of noncycling cells' factors
  noncycling_cells<-seurat_meta_toUse[!(seurat_meta_toUse$signature_1 %in% cyclingFactor),]
  noncycling_classify<- noncycling_cells$signature_1
  names(noncycling_classify)<- rownames(noncycling_cells)

  ## Merge cycling/noncycling together
  all_cycling_reclassify<-c(cycling_reclassify, noncycling_classify )

cc_reclassify_colors<- gg_colors_all[names(gg_colors_all) %in% unique(noncycling_classify)]
cc_reclassify_numbers<- sort(unique(all_cycling_reclassify))
cell.colors.reclassifycc<- as.character(plyr::mapvalues(x = all_cycling_reclassify, 
                                                        from =cc_reclassify_numbers, to = cc_reclassify_colors))
names(cell.colors.reclassifycc)<- names(all_cycling_reclassify)

## Show colors of each program
color_df<- data.frame(cluster_numbers, gg_colors)
group.colors <- gg_colors; names(group.colors)<- cluster_numbers

ggplot(color_df ,aes(x=cluster_numbers, fill=cluster_numbers)) +
  geom_bar()+
  scale_fill_manual(values=group.colors)
```



# Filtering genes to leave those that exceed some pre-defined g to the average expression magnitude
```{r}
# filter expression matrices based on some minimum max-cluster averages
emat <- filter.genes.by.cluster.expression(emat,cell.colors,min.max.cluster.average = 5)
nmat <- filter.genes.by.cluster.expression(nmat,cell.colors,min.max.cluster.average = 1)
smat <- filter.genes.by.cluster.expression(smat,cell.colors,min.max.cluster.average = 0.5)
# look at the resulting gene set
length(intersect(rownames(emat),rownames(nmat)))
```

## Save raw data for velocyto analysis 
```{r}
velocyto_obj = list()
velocyto_obj[["emat"]] = emat
velocyto_obj[["nmat"]] = nmat
velocyto_obj[["smat"]] = smat
velocyto_obj[["emb"]] = emb
velocyto_obj[["nmf_colors"]] = cell.colors
velocyto_obj[["sample_colors"]] = cell.colors.samples
velocyto_obj[["cell_names"]] = cell_names
save(velocyto_obj, file=paste0(velocyto_analysis_folder, "velocyto_obj.Robj"))
```

## The most robust estimate, that combines cell kNN pooling with the gamma fit based on an extreme quantiles:
```{r}
fit.quantile <- 0.05
rvel.qf <- gene.relative.velocity.estimates(emat,nmat,
                                            deltaT=1,
                                            kCells = 5,
                                            fit.quantile = fit.quantile)
save(rvel.qf, file=paste0(velocyto_analysis_folder, "cellkNN_gammafit_BT1478_Factors9.10.14.Robj"))

jpeg(filename=paste0(velocyto_fig_folder, "PCA_velocity_cellKNN.png"), width = 1200, height = 800)
pca.velocity.plot(rvel.qf,
                  nPcs=5,
                  plot.cols=2,
                  cell.colors=cell.colors,
                  cex=1.2,
                  pcount=0.1,
                  arrow.scale=0.2,
                  pc.multipliers=c(1,-1,-1,-1,-1))
dev.off()

jpeg(filename=paste0(velocyto_fig_folder, "PCA_velocity_cellKNN_2PCs_BT1478_Factors9.10.14_LIGER.png"), 
     width = 1200, height = 800)
pca.velocity.plot(rvel.qf,
                  nPcs=2,
                  plot.cols=1,
                  cell.colors=ac(cell.colors),
                  cex=1.2,
                  pcount=0.1,
                  arrow.scale=0.2,
                  pc.multipliers=c(1,-1))
dev.off()
```

## Compute diffusion map embedding 
```{r}
## Compute diffusion map embedding  
cm_norm = as.matrix(seurat_obj@assays$RNA@counts[seurat_obj@assays$RNA@var.features,][,cells_use])
rd_dest = DiffusionMap(t(cm_norm))
dest_embed = NULL
for (i in 1:2){
  name=paste0("DC", i)
  dest_embed = cbind(dest_embed, rd_dest[[name]])
}
colnames(dest_embed) = c("DC1", "DC2")
rownames(dest_embed) = cell_names
```


## Visulization on other embeddings (e.g. tSNE, diffusion map)
```{r}
vel <- rvel.qf
cell.colors = cell.colors
arrow.scale=1.5
cell.alpha=0.4
cell.cex=1
width=4.5
height=4

## diffusion map
jpeg(filename=paste0(velocyto_fig_folder, "UMAP_velocyto.jpg"), width = 1600, height = 1200, res = 300)
show.velocity.on.embedding.cor(dest_embed, 
                               vel, 
                               n=100, 
                               scale='sqrt',
                               cell.colors=ac(cell.colors,alpha=cell.alpha),
                               cex=cell.cex,
                               arrow.scale=arrow.scale,
                               arrow.lwd=1)
dev.off()

## UMAP
jpeg(filename=paste0(velocyto_fig_folder, "UMAP_velocyto_MUV6566_MUV6566UMAP.jpg"), width = 1600, height = 1200, res = 300)
show.velocity.on.embedding.cor(emb=emb, 
                               vel=vel, 
                               n=100, 
                               scale='sqrt',
                               cell.colors=ac(cell.colors),
                               cex=cell.cex,
                               arrow.scale=arrow.scale,
                               arrow.lwd=1)
dev.off()

## Harmony
jpeg(filename=paste0(velocyto_fig_folder, "Harmony_velocyto_NoBT1478.jpg"), 
     width = 1600, height = 1200, res = 300)
show.velocity.on.embedding.cor(emb=emb_harmony, 
                               vel=vel, 
                               n=100, 
                               scale='sqrt',
                               cell.colors=ac(cell.colors),
                               cex=cell.cex,
                               arrow.scale=arrow.scale,
                               arrow.lwd=1)
dev.off()

```


## Fitting of individual genes
```{r}
# define custom pallet for expression magnitude
neuronal_gene= "STMN2"
undiff_gene = "MAFF"
diff_gene = "BCAN"
ependymal_gene = "DNAAF1"
##astrocytic_gene = "ALDOC"

jpeg(filename=paste0(velocyto_fig_folder, "FittingOfIndvGenes.png"), width = 1200, height = 800)
gene.relative.velocity.estimates(emat,nmat, kCells = 5,
                                 fit.quantile =fit.quantile,
                                 old.fit=rvel.qf,show.gene=undiff_gene,
                                 cell.emb=emb,cell.colors=cell.colors)
dev.off()
```

## Using spanning reads to fit the gene offsets
```{r}
rvel <- gene.relative.velocity.estimates(emat,nmat,smat=smat, kCells = 5, fit.quantile=fit.quantile, diagonal.quantiles = TRUE)

jpeg(filename=paste0(velocyto_fig_folder, "PCA_SpanningReadsFitGeneOffsets.png"), width = 1200, height = 800)
pca.velocity.plot(rvel,nPcs=5,plot.cols=2,cell.colors=ac(cell.colors,alpha=0.7),
                  cex=1.2,pcount=0.1,pc.multipliers=c(1,-1,1,1,1))
dev.off()
```

## Most basic version of velocity estimates, using relative gamma fit, without cell kNN smoothing (i.e. actual single-cell velocity)
```{r}
rvel1 <- gene.relative.velocity.estimates(emat,nmat,deltaT=1,deltaT2 = 1,
                                          kCells = 1, fit.quantile=fit.quantile)

jpeg(filename=paste0(velocyto_fig_folder, "PCA_velocity_RelativeGammaFit.png"), width = 1200, height = 800)
pca.velocity.plot(rvel1,nPcs=5,plot.cols=2,cell.colors=ac(cell.colors,alpha=0.7),
                  cex=1.2,pcount=0.1,pc.multipliers=c(1,-1,1,1,1))
dev.off()
```



## Velocity vector field
```{r}
vel <- rvel.qf; cell.colors = cell.colors
arrow.scale=1.5; cell.alpha=0.4; cell.cex=1; width=4.5; height=4

## PC embedding here
pca_emb_list = pca.velocity.plot(rvel.qf,
                  nPcs=2,
                  plot.cols=1,
                  cex=1.2,
                  pcount=0.1,
                  arrow.scale=0.2,
                  pc.multipliers=c(1,-1))
pca_emb = pca_emb_list$epc@scores
saveRDS(pca_emb, paste0(velocyto_analysis_folder, "pca_embedding.rds"))

jpeg(paste0(velocyto_fig_folder, "velocyto_vector_field.jpg"), width=2400, height=1800, res=300)
show.velocity.on.embedding.cor(pca_emb,vel,n=100,scale='sqrt',cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,arrow.scale=arrow.scale,show.grid.flow=TRUE,min.grid.cell.mass=0.5,grid.n=20,arrow.lwd=2)
dev.off()
```

## Final plot
```{r}
jpeg(paste0(velocyto_fig_folder, "velocyto_metagene.jpg"), width=2400, height=1800, res=300)
pca.velocity.plot(rvel.qf,
                  nPcs=2,
                  plot.cols=1,
                  cell.colors=ac(cell.colors,alpha=0.75),
                  cex=0.6,
                  pcount=0.1,
                  arrow.scale=0.1,
                  pc.multipliers=c(1,-1))
dev.off()
```

## Save velocyto results
```{r}
velocyto_res_list = list()
velocyto_res_list[["qf"]] = rvel.qf
velocyto_res_list[["spanning"]] = rvel
save(velocyto_res_list, file=paste0(velocyto_analysis_folder, "velocyto_res_list.Robj"))
```

## Reload all data from save objects
```{r}
load("analysis/velocyto_obj.Robj")
load("analysis/velocyto_res_list.Robj")

cell_names = velocyto_obj$cell_names
sample_colors = velocyto_obj$sample_colors
subtype_colors = velocyto_obj$subtype_colors
cc_colors = velocyto_obj$cc_colors
nmf_colors = velocyto_obj$nmf_colors

emat = velocyto_obj$emat
nmat = velocyto_obj$nmat
smat = velocyto_obj$smat

rvel.qf = velocyto_res_list$qf
```