---
title: "PostTracer_Processing"
author: "Orr Ashenberg"
date: "11/04/21"
output: html_document
---
This script analyzes the T cells from glioblastoma samples. It takes the T cells clustered by the script ss2_pedbrain_cluster.Rmd, and loads that Seurat object. The T cells have already had quality control filtering for genes and cells, and clustering performed. Here, we can do follow up TCR and TraCeR analysis.

Run the first two chunks of this markdown file, before running any other chunks.

```{r init}
# Load libraries
rm(list = ls())
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
library(Seurat)
library(dplyr)
library(Matrix)
library(GSA)
library(ggbeeswarm)

# Set directories
user.path <- "/Users/jlabelle/"
proj.path<- paste0(user.path, 
                     "Dropbox (Partners HealthCare)/FilbinLab/",
                     "data_analysis/Projects/Immune/PD1_Treat/Immune/")

name.project <- "PD1"


# Load T cell Seurat object generated by ss2_pedbrain_cluster.Rmd.
# load("/Users/orr/Documents/Rdata/2020_12_10_pedbrain_immune_no_batchcorrect.Rda") 
Rda.nobatchcorrect.path <- paste0(proj.path, "analysis/detailed_annotation/HarmonyInt/tcell_seurat.Rds")
gcdata<- readRDS(Rda.nobatchcorrect.path)
gcdata$sampleid<- gsub("MUV0", "MUV", gcdata$sampleid)

# Set project variables using annotation.
experiment <- factor(unique(gcdata$sampleid), levels = unique(gcdata$sampleid)) 
sampleid <- unique(gcdata$sampleid)

# Settings for T cell types.
Tcelltype <- "CD8"
types <- list("CD4" = c("CD4 T"), "Treg" = c("Treg"), "CD4_Treg" = c("CD4 T",  "Treg"), "CD8" = c("CD8 T"))
cell.type <- types[[Tcelltype]]

# Order in which to display annotated T cell subsets when plotting.
group.order <- c("CD8 T", "CD4 T", "Treg", "Cycling T cells", "Mono")

# Set minimum size of clonotype to consider.
group_size.cutoff <- 2  # need at least this many cells to define a clonotype
# Number of iterations of choosing random sets of cells in nonclonal gene expression analysis.
# n.runs <- 5
n.runs <- 10

# Output figures and Broad single-cell portal
figures.dir <- paste0(proj.path, "/figures/tracer/")    
figures.path <- sapply(sampleid, function(i) paste0(figures.dir, i, "_"), USE.NAMES = F) 
#portal.path <- paste0(figures.dir, "/portal")
Rda.clonalTcells.path<- paste0("analysis/tracer/clonalTcells_cluster.Rda")

# Save seurat object expression matrix data, tSNE, clustering, and cluster annotation scores.
Rda.clonalTcells.path <- paste0(proj.path, "/analysis/tracer/clonalTcells_cluster.Rda")
#Rda.clonalTcells.NMF.path <- paste0(proj.path, "/src/Rdata/", date, "_clonalTcells_NMF_cluster.Rda")

# Read in a list of cell cycle markers, from Tirosh et al, 2015.
# We can segregate this list into markers of G2/M phase and markers of S phase.
cc.genes <- read.table(paste0(user.path,"Dropbox (Partners HealthCare)/FilbinLab/data_analysis/",
                              "Marker_genes/regev_lab_cell_cycle_genes.txt"))$V1
s.genes <- cc.genes[1:43]
g2m.genes <- cc.genes[44:97]


# Load my functions.
source(paste0(user.path, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/", "/Scripts/plotutils.R"))
source(paste0(user.path, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/", "/Scripts/seurat3utils.R"))
source(paste0(user.path, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/", "/Scripts/color.R"))
```



## TraCeR T-cell receptor reconstruction statistics
We look at number of total cells (some cells may be missing TCR reads), number of productive A chains, productive B chains, number of productive A and B chains, number of expanded T cells in clonotypes, and number of clonotypes.
```{r tracer_stats}
run.exps <- as.character(sampleid)

# List of clonally expanded T cells keyed by experiment patient sample that the cells come from.
# clonal is list containing the name of the clonotype cells in each experiment.
# clonotypes.cells is list containing all the metadata for each clonotype cell in each experiment.
clonal <- vector("list", length = length(run.exps))
clonotypes.cells <- vector("list", length = length(run.exps))  # each experiment is a list of dataframes, and each dataframe is a clonal group of cells from that experiment
names(clonal) <- run.exps
names(clonotypes.cells) <- run.exps 

# Make a dataframe containing cells from all clonotypes across all experiments that pass quality control. 
# Clonal groups get renumbered to ensure unique clonotype group numbers across all experiments.
clonotypes.all <- data.frame(cell_name = character(), experiment = character(), clonal_group = vector(), group_size = vector(), stringsAsFactors = F)
group.offset <- 0  # offset gets updated after each experiment to ensure unique group number for clonotypes

# Store a vector of number of T cells found in each clonotype.
clonotypes.size <- vector()  

# Store a list where each element is keyed by experiment and contains a vector of the names of cells 
# that have a reconstructed TCR.
tcell.reconstructed <- vector("list", length = length(run.exps))

# Experiment id, number of cells passing QC, number of cells with productive A or B chain, 
# number of cells with productive A and B chains, 
# number of clonally expanded T cells, number of T cell clonotypes
df.plot <- data.frame(experiment = character(), nExps = vector(), nCells = vector(), productiveAB = vector(), 
                      nExpanded = vector(), nClonotypes = vector())
run.exps<- c("BT1910", "BT1935")

for (exp in run.exps) {
  exp.filename <- exp
  print(exp)
  file.celldata <- paste0(proj.path, "/data/tracer_output/", exp.filename, "/cell_data.csv")
  
  # If cell_data file doesn't exist, that means no TCR chains were reconstructed. 
  # Collect basic stats and skip to next sample
  if (!file.exists(file.celldata)) {
    tcell.reconstructed[[exp]] <- character()
    n.exps <- filter(gcdata[[]], sampleid == exp) %>% nrow()  # T cells in this experiment passing quality control
    n.cells <- 0  # cells in this experiment with productive A chain or productive B chain
    n.productiveAB <- 0  # productive A chain and productive B chain
    n.expanded <- 0  # expanded T cells
    n.clonotypes <- 0  # clonotypes
    df.plot <- rbind(df.plot, data.frame(exp, n.exps, n.cells, n.productiveAB, n.expanded, n.clonotypes))
    next 
  }
  
  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(file.celldata, header = T, sep = ",", stringsAsFactors = F)
  
  # Rename cell names as needed to match gcdata
  tcr.annotate$cell_name<- gsub("_", "\\.", gsub("highdepth.", "", tcr.annotate$cell_name))

  # Identify all T cells in this experiment containing a reconstructed TCR chain A or chain B.
  tcell.reconstructed[[exp]] <- tcr.annotate[!(tcr.annotate$A_productive=="") | !(tcr.annotate$B_productive==""), "cell_name"]
    
  # Identify clonally expanded T cells.
  if (!all(is.na(tcr.annotate$clonal_group))) {  # fix this line
    label.clonotypes <- c(0:max(tcr.annotate$clonal_group, na.rm = T))  # clonotype number labels start at 0
    clonotypes <- filter(tcr.annotate, clonal_group %in% label.clonotypes)  # T cells that are clonally expanded
  } else {
    clonotypes <- data.frame()
  }
  
  # Collect stats on clonally expanded T cells.
  n.exps <- filter(gcdata[[]], sampleid == exp) %>% nrow()  # T cells in this experiment passing quality control
  n.cells <- tcr.annotate[!(tcr.annotate$A_productive=="") | !(tcr.annotate$B_productive==""),] %>% nrow  # cells in this experiment with productive A chain or productive B chain
  n.productiveAB <- tcr.annotate[!(tcr.annotate$A_productive=="") & !(tcr.annotate$B_productive==""),] %>% nrow  # productive A chain and productive B chain
  n.expanded <- nrow(clonotypes)  # expanded T cells
  n.clonotypes <- length(unique(clonotypes$clonal_group))  # clonotypes
  df.plot <- rbind(df.plot, data.frame(exp, n.exps, n.cells, n.productiveAB, n.expanded, n.clonotypes))
  
  # Keep clonally expanded T cells that are also found in Tcell gene expression matrix and not filtered out.
  # Names in gene expression matrix should match those in TraCeR data (E30TC_P5_E12_E30_CD4+CD8+_P5P6).
  clonal[[exp]] <- intersect(clonotypes$cell_name, colnames(gcdata))
  remove.cells <- setdiff(clonotypes$cell_name, clonal[[exp]])
  clonotypes <- clonotypes[!clonotypes$cell_name %in% remove.cells, ]
  if (nrow(clonotypes)) {
    clonotypes <- clonotypes %>% group_by(clonal_group) %>% mutate(group_size = length(clonal_group))  # update number of cells in each clonotype after having removed cells that were not found in expression matrix
    clonotypes <- clonotypes[clonotypes$group_size > 1, ]  # remove clonotypes that now only have 1 cell
  }
  print(c(length(clonal[[exp]]), n.expanded))
  if (length(remove.cells)) {print(remove.cells)}  # cells that were likely filtered from expression matrix

  # For each clonotype, store the cells that make up that clonotype.  
  clonotypes.cells[[exp]] <- split(clonotypes, clonotypes$clonal_group)

  # Store size of each clonotype.
  size <- sapply(clonotypes.cells[[exp]], function(clone) unique(clone$group_size)) %>% unname()
  if (length(size)) {
    clonotypes.size <- c(clonotypes.size, size)
    print(paste0("clonotype sizes: ", paste(size, collapse = " ")))
  }
  
  # Place cells for all clonotypes across all experiments in a single dataframe, with each clonotype 
  # group receiving a unique number. Within each experiment, clonal_group numbering starts at 0.
  if (nrow(clonotypes)) {
    clonotypes <- clonotypes[order(clonotypes$clonal_group),]  # sort by clonal_group
    clonotypes$clonal_group <- clonotypes$clonal_group + group.offset
    clonotypes[["experiment"]] <- exp 
    clonotypes$cell_name <- as.character(clonotypes$cell_name)  # remove factor levels from cell names
    clonotypes.all <- bind_rows(clonotypes.all, clonotypes[, c("cell_name", "experiment", "clonal_group", "group_size")])
    group.offset <- max(clonotypes$clonal_group) + 1  # update clonal_group number for next experiment in iteration of this loop
  }
}

```

## Plot clonotype/cell reconstruction results
```{r}
# Set order of experiments when plotting
df.plot$exp <- factor(df.plot$exp, levels = sampleid)

# Gather df.plot dataframe into stat : count mapping key-value pairs for making bar plots. 
labels.plot <- c("number of cells", "productive A or B chain", "productive A and B chain", "T cells with recurrent TCRs", "clonotypes")
bar.plot <- gather(df.plot, "stat", 'count', c(2:6))
# Set order in which different statistics from TraCer are plotted.
bar.plot$stat <- factor(bar.plot$stat, levels = colnames(df.plot)[2:6])
ggplot(data = bar.plot) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = position_dodge()) + 
  labs(x = "experiment", y = "count") + 
  theme(axis.text.x  = element_text(angle=90)) 

ggsave(paste0(figures.dir, 'tracer_stats.png'), width = 6, height = 6)

# Graph number of cells passing QC and number of productive T cells per patient
labels.plot <- c("Total T cells", bquote("TCR" ~ alpha ~ "or" ~ beta))
bar.plot <- gather(df.plot, "stat", 'count', c(2:3))
# Set order in which different statistics from TraCer are plotted.
bar.plot$stat <- factor(bar.plot$stat, levels = colnames(df.plot)[2:3])
# ggplot(data = bar.plot) + geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = position_dodge()) + scale_fill_manual(values = material.heat(2), name="", breaks=levels(bar.plot$stat), labels=labels.plot) + labs(x = "experiment", y = "Number of cells") + theme(axis.text.x  = element_text(angle=90)) 
ggplot(data = bar.plot) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = "identity") + 
  labs(x = "Sample", y = "Number of cells") + theme(axis.text.x  = element_text(angle=90)) 
ggsave(paste0(figures.dir, 'fig2_tracer_ncells_nproductivechain.png'), width = 8, height = 6)

# Graph number of clonotypes per patient.
ggplot(data = df.plot[, c("exp", "n.clonotypes")]) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = n.clonotypes)) + 
  labs(x = "experiment", y = "Number of clonotypes") + 
  theme(axis.text.x  = element_text(angle=90)) 
ggsave(paste0(figures.dir, 'fig2_tracer_nclonotypes.png'), width = 6, height = 6)

# Graph percent of T cells with productive A or B chain that are clonally expanded in each patient.
df.plot$percent.expanded <- 100 * df.plot$n.expanded / df.plot$n.cells
ggplot(data = df.plot[, c("exp", "percent.expanded")]) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = percent.expanded)) + 
  labs(x = "experiment", y = "T cells with recurrent TCRs (%)") + 
  theme(axis.text.x  = element_text(angle=90)) + ylim(c(0, 50))
ggsave(paste0(figures.dir, 'fig2_tracer_expanded.png'), width = 6, height = 6)

# Scatter plot of T cells with recurrent TCRS (%) vs Number of clonotypes.
ggplot(data = df.plot, mapping = aes(x = n.clonotypes, y = percent.expanded, label = exp)) + 
  geom_point(mapping = aes(size = n.cells)) + 
  geom_text(nudge_y = 2) + 
  labs(x = "Number of clonotypes", y = "T cells with recurrent TCRs (%)") + 
  scale_size_continuous(name = bquote(atop("Number of TCR", ~ alpha ~ "or" ~ beta))) + 
  ylim(0, 52)
ggsave(paste0(figures.dir, 'fig2_tracer_expanded_vs_nproductivechain.png'), width = 8, height = 6)
cor(x = df.plot$n.clonotypes, y = df.plot$percent.expanded, method = "pearson")

# Distribution of clonotype sizes in each experiment. 
# These plots show both the clonotype sizes, and the sizes of the clonotypes normalized by the 
# number of cells in that experiment for which there was a productive A chain or B chain.
df.size <- clonotypes.all %>% group_by(experiment, clonal_group) %>% summarise(size = unique(group_size))
df.size$experiment <- factor(df.size$experiment, levels = levels(experiment))
df.size$size.normalized <- 0  # normalize clonotype size by number of cells
for (exp in unique(experiment)) {
  df.size[df.size$experiment == exp, "size.normalized"] <- 100 * df.size[df.size$experiment == exp, "size"] / 
    df.plot[df.plot$exp == exp, "n.cells"]
}
ggplot(data = df.size, mapping = aes(x = experiment, y = size)) + 
  geom_quasirandom(groupOnX = T, cex = 1, varwidth = T) + 
  labs(y = "clonotype size") + scale_y_continuous(trans = "log2")
ggsave(paste0(figures.dir, "fig2_tracer_clonotype_size_beeswarm.png"), width = 15, height = 6)


ggplot(data = df.size, mapping = aes(x = experiment, y = size.normalized)) + 
  geom_quasirandom(groupOnX = T, cex = 1, varwidth = T) + 
  labs(y = "percent of reconstructed T cells in clonotype")
ggsave(paste0(figures.dir, "tracer_clonotype_sizenormalized_beeswarm.png"), width = 15, height = 6)

# Histogram of clonotype sizes (number of T cells in a clonotype).
ggplot(data = data.frame(size = clonotypes.size)) + 
  geom_histogram(mapping = aes(x = size), binwidth = 1, color="black", fill="white") + 
  scale_x_continuous(breaks=seq(0,60,5)) + 
  scale_y_continuous(expand = c(0,0)) + 
  labs(x = "clonotype size", y = "count") 
ggsave(paste0(figures.dir, 'tracer_clonotype_histogram.png'), width = 8, height = 6)

# Write a file summarizing all clonotypes.
rownames(clonotypes.all) <- clonotypes.all$cell_name
write.csv(clonotypes.all, file=paste0(analysis.dir, "clonotypes_all.csv"), quote = F)

# Save current progress.
save( clonal, clonotypes.cells, clonotypes.all, tcell.reconstructed, file = Rda.clonalTcells.path)

# Write expression matrix for Itay.
#write.table(as.matrix(gcdata@data), file = paste0(figures.dir, 'expression_matrix.tsv'), sep = '\t', quote = F)
```


## The rest of this script is not run
## Instead, merge the results (saved above) with the rest of the samples
```{r}
new_samples_list<-list(gcdata=gcdata, clonal=clonal, clonotypes.cells=clonotypes.cells, 
                       clonotypes.all=clonotypes.all, tcell.reconstructed=tcell.reconstructed)

## Load in results (from Orr) for the other samples
load("../data/2020_12_10_clonalTcells_cluster.rda")
original_samples_list<-list(gcdata=gcdata, clonal=clonal, clonotypes.cells=clonotypes.cells, 
                       clonotypes.all=clonotypes.all, tcell.reconstructed=tcell.reconstructed)

merged_samples_list<-list()  

## gcdata is already merged (NewImmuneSamples_IntWithOriginal). Load in
load("analysis/gcdata_original_new_merge.Robj")
merged_samples_list$gcdata<- gcdata_merge

## Add clonal information to merged. There are no clonal groups for the new samples, so will all be nonclonal/na

## Concatenate all other objects to save
merged_samples_list$clonal<- c(new_samples_list$clonal, original_samples_list$clonal)
merged_samples_list$clonotypes.cells<- c(new_samples_list$clonotypes.cells, original_samples_list$clonotypes.cells)
merged_samples_list$clonotypes.all<- rbind(new_samples_list$clonotypes.all, original_samples_list$clonotypes.all)
merged_samples_list$tcell.reconstructed<- c(new_samples_list$tcell.reconstructed, original_samples_list$tcell.reconstructed)


gcdata <- merged_samples_list$gcdata
clonal<- merged_samples_list$clonal
clonotypes.cells<- merged_samples_list$clonotypes.cells
clonotypes.all <- merged_samples_list$clonotypes.all
tcell.reconstructed<- merged_samples_list$tcell.reconstructed

## save
save(gcdata, 
     clonal, 
     clonotypes.cells, 
     clonotypes.all, 
     tcell.reconstructed, file = paste0("analysis/2021_09_01_clonalTcells_cluster.Rda"))
```



## Store which cells are clonally expanded in the Seurat object and compare differential expression in clonally expanded vs non-clonally expanded cells
I identified clonal cells at the top of this script, and now we will compare gene expression in clonal vs non-clonal cells. This is broken up into the different T cell types. In later sections, across clonotypes, we compare highly expanded and lowly expanded clonotypes.
```{r DE_clonal}
# Possible DE tests to run in this code chunk.
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]

# Cluster of cycling T cells has CD4+ and CD8+ cells. Divide into CD4+ or CD8+ based on expression of 
# one gene and lack of expression of other gene. Caveat: Dropouts could result in CD4 or CD8 not 
# being detected, but this is a small number of cells.
# cycling <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate == "Cycling T cells",])
# df.type <- data.frame("CD4" = gcdata.sub@data["CD4", cycling], "CD8" = Matrix::colMeans(gcdata.sub@data[c("CD8A", "CD8B"), cycling]))
# ggplot(data = df.type) + geom_point(mapping = aes(x = CD4, y = CD8))
# CD4.cycling <- rownames(df.type[df.type$CD4 > 1 & df.type$CD8 < 1, ])
# CD8.cycling <- rownames(df.type[df.type$CD8 > 1 & df.type$CD4 < 1, ])
# unname(gcdata.sub@data["CD4", CD4.cycling])
# unname(gcdata.sub@data["CD8A", CD8.cycling])

title = "All T cells"
cells.clonal <- unlist(clonal)
cells.nonclonal <- setdiff(colnames(gcdata), cells.clonal)
# Within all T cells, clonal is class 1, non-clonal is class 2.
gcdata[["clonal"]] <- 0
gcdata$clonal[cells.clonal] <- "clonal"
gcdata$clonal[cells.nonclonal] <- "not clonal"
DimPlot(gcdata, reduction = "umap", group.by = "clonal", cols = material.heat(n_distinct(gcdata$clonal)), label = F, pt.size = 1)
ggsave(paste0(figures.dir, 'UMAP_clonotype_category.png'), width = 6, height = 6)

title = "CD4+ T cells"  # exclude Tregs
cells.subset <- c(rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate %in% c("CD4+ (P1)", "CD4+ (P2)"), ]), CD4.cycling)
cells.clonal <- unlist(clonal)
cells.clonal <- intersect(cells.subset, cells.clonal)
cells.nonclonal <- setdiff(cells.subset, cells.clonal)
# Within CD4+, clonal is class 1, non-clonal is class 2. Non-CD4+ is class 0.
gcdata.sub@meta.data["clonal.CD4"] <- 0
gcdata.sub@meta.data[cells.clonal, "clonal.CD4"] <- 1
gcdata.sub@meta.data[cells.nonclonal, "clonal.CD4"] <- 2

title = "CD8+ T cells"
cells.subset <- c(rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate %in% c("CD8+"), ]), CD8.cycling)
cells.clonal <- unlist(clonal)
cells.clonal <- intersect(cells.subset, cells.clonal)
cells.nonclonal <- setdiff(cells.subset, cells.clonal)
# Within CD8+, clonal is class 1, non-clonal is class 2. Non-CD8+ is class 0.
gcdata.sub@meta.data["clonal.CD8"] <- 0
gcdata.sub@meta.data[cells.clonal, "clonal.CD8"] <- 1
gcdata.sub@meta.data[cells.nonclonal, "clonal.CD8"] <- 2

title = "Treg cells"
cells.subset <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate == "Treg", ])
cells.clonal <- unlist(clonal)
cells.clonal <- intersect(cells.subset, cells.clonal)
cells.nonclonal <- setdiff(cells.subset, cells.clonal)
# Within Treg, clonal is class 1, non-clonal is class 2. Non-Treg is class 0.
gcdata.sub@meta.data["clonal.Treg"] <- 0
gcdata.sub@meta.data[cells.clonal, "clonal.Treg"] <- 1
gcdata.sub@meta.data[cells.nonclonal, "clonal.Treg"] <- 2

# Get fraction of clonally expanded T cells in each cluster.
# df.clonal <- as.data.frame(prop.table(table(gcdata.sub@meta.data$clonal, gcdata.sub@meta.data$res.1), 2))
df.clonal <- as.data.frame(prop.table(table(gcdata.sub@meta.data$clonal, gcdata.sub@meta.data$annotate), 2))
colnames(df.clonal) <- c("clonal", "cluster", "percent")
# df.clonal$cluster <- factor(df.clonal$cluster, levels = as.character(order(as.numeric(df.clonal$cluster))))
df.clonal$clonal <- plyr::mapvalues(x = df.clonal$clonal, from = c(1, 2), to = c("clonal", "not clonal"))
df.clonal$clonal <- factor(df.clonal$clonal, levels = c("not clonal", "clonal"))
df.clonal$percent <- 100 * df.clonal$percent
ggplot(data = df.clonal) + geom_bar(stat = "identity", mapping = aes(x = cluster, y = percent, fill = clonal)) + scale_fill_manual(values = material.heat(2)) 
ggsave(paste0(figures.dir, 'tracer_percent_clonal_cluster.png'), width = 4, height = 5)
df.clonal <- df.clonal %>% filter(clonal == "clonal")
ggplot(data = df.clonal[, c("cluster", "percent")]) + geom_bar(stat = "identity", mapping = aes(x = cluster, y = percent)) + scale_fill_manual(values = material.heat(1)) + labs(x = "cluster", y = "T cells with recurrent TCRs (%)") + theme(axis.text.x  = element_text(angle=90)) + ylim(c(0, 50))
ggsave(paste0(figures.dir, 'fig2_tracer_percent_clonal_cluster.png'), width = 4, height = 6)
ggsave(paste0(figures.dir, 'fig2_tracer_percent_clonal_cluster.pdf'), width = 4, height = 6)
freq.clonal <- prop.table(table(gcdata.sub@meta.data$clonal)) # overall proportion of clonal and non-clonal T cells.

# Differential expression test for clonal vs non-clonal CD4+ T cells (exclude Treg).
gcdata.sub <- SetAllIdent(gcdata.sub, id = "clonal.CD4")
#gcdata.markers <- FindMarkers(gcdata, ident.1 = 1, ident.2 = 2, min.pct = 0.2, logfc.threshold = 0.5, print.bar = F)
gcdata.markers <- FindMarkers(gcdata.sub, ident.1 = 1, ident.2 = 2, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.sub@data>0) > 0.2*length(gcdata.sub@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_clonal_CD4.csv"))
cells <- rownames(gcdata.sub@meta.data[(gcdata.sub@meta.data$clonal.CD4 == 1) | (gcdata.sub@meta.data$clonal.CD4 == 2), ])
DoHeatmap(gcdata.sub, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, 
          group.by = "clonal.CD4", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T) + 
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
ggsave(paste0(figures.dir, "top_DE_genes_clonal_CD4_expression.png"), width = 4, height = 16)

# Differential expression test for clonal vs non-clonal CD8+ T cells.
gcdata.sub <- SetAllIdent(gcdata.sub, id = "clonal.CD8")
gcdata.markers <- FindMarkers(gcdata.sub, ident.1 = 1, ident.2 = 2, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.sub@data>0) > 0.2*length(gcdata.sub@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_clonal_CD8.csv"))
cells <- rownames(gcdata.sub@meta.data[(gcdata.sub@meta.data$clonal.CD8 == 1) | (gcdata.sub@meta.data$clonal.CD8 == 2), ])
DoHeatmap(gcdata.sub, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, 
          group.by = "clonal.CD8", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T)  + 
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
ggsave(paste0(figures.dir, "top_DE_genes_clonal_CD8_expression.png"), width = 8, height = 4)

# Differential expression test for clonal vs non-clonal Treg.
gcdata.sub <- SetAllIdent(gcdata.sub, id = "clonal.Treg")
gcdata.markers <- FindMarkers(gcdata.sub, ident.1 = 1, ident.2 = 2, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.sub@data>0) > 0.2*length(gcdata.sub@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_clonal_Treg.csv"))
cells <- rownames(gcdata.sub@meta.data[(gcdata.sub@meta.data$clonal.Treg == 1) | (gcdata.sub@meta.data$clonal.Treg == 2), ])
DoHeatmap(gcdata.sub, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, 
          group.by = "clonal.Treg", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T)  + 
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
ggsave(paste0(figures.dir, "top_DE_genes_clonal_Treg_expression.png"), width = 8, height = 4)

gcdata.sub <- SetAllIdent(gcdata.sub, id = "res.1")

# Save current progress.
save(gcdata, clonal, clonotypes.cells, clonotypes.all, tcell.reconstructed, file = Rda.clonalTcells.path)
```

## Compare clonotypes with high expansion and low expansion across all individuals
Find the largest and smallest sized clonotypes and look for differences in gene expression. The clonotypes are pooled across all the patient samples for these comparisons.
```{r expansion}
# Possible DE tests to run in this code chunk.
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]

#################
# For all T cells, compare clonotypes with many cells to those with few cells. High vs low expansion.
#################
cells.high <- clonotypes.all[clonotypes.all$group_size >= 5, ]$cell_name
cells.low <- clonotypes.all[clonotypes.all$group_size == 2, ]$cell_name
cells <- c(cells.low, cells.high)

gcdata.markers <- FindMarkers(gcdata.sub, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
gcdata.markers <- gcdata.markers[gcdata.markers$p_val <= 0.1/nrow(gcdata.sub@data), ]  # Bonferroni correction
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_allTcells_high_vs_low_expansion.csv"))
DoHeatmap(gcdata.sub, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, group.by = "annotate",
          cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T)
ggsave(paste0(figures.dir, "top_DE_genes_allTcells_high_vs_low_expansion.png"), width = 8, height = 4, dpi = 200)

gcdata.expanded <- SubsetData(gcdata.sub, cells.use = cells)
gcdata.expanded@meta.data$expanded <- NA
gcdata.expanded@meta.data[cells.low, ]$expanded <- "low"
gcdata.expanded@meta.data[cells.high, ]$expanded <- "high"

genes <- c("KLRB1", "GZMA", "GZMB", "PRF1", "IL2", "IFNG", "TNF", "IL10", "TGFB1", "TGFB2", "PDCD1", "CTLA4", "HAVCR2", "LAG3", "TIGIT")
VlnPlot(gcdata.expanded, genes, group.by = "expanded")
ggsave(paste0(figures.dir, "vln_genes_allTcells_high_vs_low_expansion.png"), width = 10, height = 10, dpi = 200)

#################
# For all CD4+ T cells excluding Tregs, compare clonotypes with many cells to those with few cells. 
# High vs low expansion.
#################
cells.high <- clonotypes.all[clonotypes.all$group_size >= 5, ]$cell_name
cells.high <- cells.high[cells.high %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["CD4"]], ])]
cells.low <- clonotypes.all[clonotypes.all$group_size == 2, ]$cell_name
cells.low <- cells.low[cells.low %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["CD4"]], ])]
cells <- c(cells.low, cells.high)

# Make a Seurat object with a subset of the data, and with T cells labeled as low or high expansion.
gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
gcdata.Tcells@meta.data[cells.high, "expansion"] <- "high"
gcdata.Tcells@meta.data[cells.low, "expansion"] <- "low"
gcdata.Tcells@meta.data$expansion <- factor(gcdata.Tcells@meta.data$expansion, levels = c("low", "high"))

gcdata.markers <- FindMarkers(gcdata.Tcells, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.Tcells@data>0) > 0.2*length(gcdata.Tcells@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_CD4_high_vs_low_expansion.csv"))
DoHeatmap(gcdata.Tcells, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, group.by = "expansion", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T) + scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
ggsave(paste0(figures.dir, "top_DE_genes_CD4_high_vs_low_expansion.png"), width = 6, height = 4, dpi = 200)

#################
# For all CD8+ T cells, compare clonotypes with many cells to those with few cells. High vs low expansion.
#################
cells.high <- clonotypes.all[clonotypes.all$group_size >= 5, ]$cell_name
cells.high <- cells.high[cells.high %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["CD8"]], ])]
cells.low <- clonotypes.all[clonotypes.all$group_size == 2, ]$cell_name
cells.low <- cells.low[cells.low %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["CD8"]], ])]
cells <- c(cells.low, cells.high)

# Make a Seurat object with a subset of the data, and with T cells labeled as low or high expansion.
gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
gcdata.Tcells@meta.data[cells.high, "expansion"] <- "high"
gcdata.Tcells@meta.data[cells.low, "expansion"] <- "low"
gcdata.Tcells@meta.data$expansion <- factor(gcdata.Tcells@meta.data$expansion, levels = c("low", "high"))

gcdata.markers <- FindMarkers(gcdata.Tcells, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.Tcells@data>0) > 0.2*length(gcdata.Tcells@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_CD8_high_vs_low_expansion.csv"))
# DoHeatmap(gcdata.Tcells, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, group.by = "expansion", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T) + scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
# ggsave(paste0(figures.dir, "top_DE_genes_CD8_high_vs_low_expansion.png"), width = 8, height = 12, dpi = 200)

#################
# For all Treg cells, compare clonotypes with many cells to those with few cells. High vs low expansion.
#################
cells.high <- clonotypes.all[clonotypes.all$group_size >= 5, ]$cell_name
cells.high <- cells.high[cells.high %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["Treg"]], ])]
cells.low <- clonotypes.all[clonotypes.all$group_size == 2, ]$cell_name
cells.low <- cells.low[cells.low %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["Treg"]], ])]
cells <- c(cells.low, cells.high)

# Make a Seurat object with a subset of the data, and with T cells labeled as low or high expansion.
gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
gcdata.Tcells@meta.data[cells.high, "expansion"] <- "high"
gcdata.Tcells@meta.data[cells.low, "expansion"] <- "low"
gcdata.Tcells@meta.data$expansion <- factor(gcdata.Tcells@meta.data$expansion, levels = c("low", "high"))

gcdata.markers <- FindMarkers(gcdata.Tcells, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.Tcells@data>0) > 0.2*length(gcdata.Tcells@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_CD8_high_vs_low_expansion.csv"))
DoHeatmap(gcdata.Tcells, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, group.by = "expansion", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T) + scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
ggsave(paste0(figures.dir, "top_DE_genes_Treg_high_vs_low_expansion.png"), width = 6, height = 4, dpi = 200)
```

## Compare clonotypes with high expansion and low expansion within an individual
Find the largest and smallest sized clonotypes within an individual, and look for differences in gene expression. The first part of this code chunk looks at all clonotypes within an individual, and the second part of this code chunk looks at CD4+ clonotypes and CD8+ clonotypes within an individual separately.
```{r expansion_individual}
sort(table(clonotypes.all$experiment))

# Possible DE tests to run in this code chunk.
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]

# Look at expression of these genes in high vs low clonal expansion.
genes <- c("CD3D", "CD4", "CD8A", "KLRB1", "GZMA", "GZMB", "PRF1", "IL2", "IFNG", "TNF", "IL10", "TGFB1", "TGFB2", "PDCD1", "CTLA4", "HAVCR2", "LAG3", "TIGIT")

# #################
# # For all T cells within individual, compare clonotypes with many cells to those with few cells. 
# # This code looks at all clonotypes, CD4+ and CD8+, together.
# # High vs low expansion.
# #################
# for (exp in c("E79", "E60", "E122", "E47", "E42", "E61", "E48")) {
#   # Find cells in clonotypes with high or low expansion.
#   cells.high <- clonotypes.all %>% filter(group_size > 2, experiment == exp)  %>% pull(cell_name)
#   cells.low <- clonotypes.all %>% filter(group_size == 2, experiment == exp)  %>% pull(cell_name)
#   cells <- clonotypes.all %>% filter(experiment == exp) %>% pull(cell_name)
#   
#   # Make a new Seurat object just with the clonotype T cells from this patient, 
#   # and mark which cells have low or high clonal expansion.
#   gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
#   gcdata.Tcells@meta.data$group_size <- clonotypes.all[rownames(gcdata.Tcells@meta.data), "group_size"]
#   
#   gcdata.markers <- FindMarkers(gcdata.Tcells, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
#   n.genes.test <- length(which(rowSums(gcdata.Tcells@data>0) > 0.2*length(gcdata.Tcells@cell.names)))
#   gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
#   gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj < 1, ]  
#   gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
#   # write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_allTcells_high_vs_low_expansion.csv"))
#   # DE genes.
#   if (nrow(gcdata.markers)) {
#     DoHeatmap(gcdata.Tcells, genes.use = rownames(gcdata.markers), slim.col.label = T, group.by = "group_size",
#             cex.row = 15, cex.col = 15, cells.use = cells, title = exp, remove.key = T)
#     ggsave(paste0(figures.dir, "top_DE_genes_", exp, "_high_vs_low_expansion.png"), width = 8, height = 4, dpi = 200)
#   }
#   # Marker genes.
#   DoHeatmap(gcdata.Tcells, genes.use = genes, slim.col.label = T, group.by = "group_size",
#             cex.row = 15, cex.col = 15, cells.use = cells, title = exp, remove.key = T)
#   ggsave(paste0(figures.dir, "marker_genes_", exp, "_high_vs_low_expansion.png"), width = 8, height = 4, dpi = 200)
# }

#################
# For all T cells within an individual, compare clonotypes with many cells to those with few cells. 
# This code divides clonotypes into CD4+ or CD8+ before analyzing them.
# High vs low expansion.
#################
for (exp in c("E79", "E60", "E122", "E47", "E42", "E61", "E48")) {
  clonotypes.CD4 <- data.frame(cell_name = character(), clonal_group = double(), group_size = integer())
  clonotypes.CD8 <- data.frame(cell_name = character(), clonal_group = double(), group_size = integer())

  # For each experiment, divide clonotypes into CD4 group or CD8 group.
  for (clone in clonotypes.cells[[exp]]) {
    group.size <- clone$group_size[1]  # number of cells in clonotype
    cells.annotate <- gcdata.sub@meta.data[as.vector(clone$cell_name), "annotate"] # annotated identity of cells
    num.cell.type.CD4 <- sum(cells.annotate %in% types[["CD4"]])  # number of cells with requested T cell type
    num.cell.type.CD8 <- sum(cells.annotate %in% types[["CD8_CD8NK"]])  # number of cells with requested T cell type
    print(paste0(clone$clonal_group[1], " ", num.cell.type.CD4, " ", num.cell.type.CD8, " ", group.size, " ", paste0(cells.annotate, collapse = " ")))

    # Mean scaled expression of CD4 or CD8 in cells from this clonotype.
    mean.CD4 <- mean(gcdata.sub@scale.data["CD4", as.character(clone$cell_name)])
    mean.CD8 <- mean(gcdata.sub@scale.data[c("CD8A", "CD8B"), as.character(clone$cell_name)])
    
    # Clonotype is called CD4 or CD8 based on majority rule and mean expression of CD4 relative to CD8.
    if ( (num.cell.type.CD4 > num.cell.type.CD8) & (mean.CD4 > mean.CD8) ) {
      clonotypes.CD4 <- rbind(clonotypes.CD4, as.data.frame(clone[, c("cell_name", "clonal_group", "group_size")]))
    } else if ( (num.cell.type.CD4 < num.cell.type.CD8) & (mean.CD4 < mean.CD8) ) {
      clonotypes.CD8 <- rbind(clonotypes.CD8, as.data.frame(clone[, c("cell_name", "clonal_group", "group_size")]))
    } else if (mean.CD4 > mean.CD8) {
      clonotypes.CD4 <- rbind(clonotypes.CD4, as.data.frame(clone[, c("cell_name", "clonal_group", "group_size")]))
    } else {
      clonotypes.CD8 <- rbind(clonotypes.CD8, as.data.frame(clone[, c("cell_name", "clonal_group", "group_size")]))
    }
  }
  
  # Order clonotypes in their dataframes by their group size. 
  # This is useful for plotting the clonotypes ordered by size.
  clonotypes.CD4 <- clonotypes.CD4 %>% arrange(group_size, clonal_group)
  clonotypes.CD8 <- clonotypes.CD8 %>% arrange(group_size, clonal_group)
  rownames(clonotypes.CD4) <- clonotypes.CD4$cell_name
  rownames(clonotypes.CD8) <- clonotypes.CD8$cell_name
  
  # Loop over CD4 clonotypes and over CD8 clonotypes.
  clonotypes.loop <- list("CD4+" = clonotypes.CD4, "CD8+" = clonotypes.CD8)
  for (i in seq(clonotypes.loop)) {
    clonotypes <- clonotypes.loop[[i]]
    name <- names(clonotypes.loop)[i]
    title <- paste0(exp, " ", name)
    
    # Find cells in clonotypes with high or low expansion.
    min.group.size <- min(clonotypes$group_size)  # get size of smallest clonotype
    cells.high <- clonotypes %>% filter(group_size > min.group.size)  %>% pull(cell_name)
    cells.low <- clonotypes %>% filter(group_size == min.group.size)  %>% pull(cell_name)
    cells <- as.character(clonotypes$cell_name)
    
    # Make a new Seurat object just with the clonotype T cells from this patient
    # and mark clonotype groups and sizes.
    gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
    gcdata.Tcells@meta.data$group_size <- clonotypes[rownames(gcdata.Tcells@meta.data), "group_size"]
    gcdata.Tcells@meta.data$clonal_group <- clonotypes[rownames(gcdata.Tcells@meta.data), "clonal_group"]

    # Marker genes.
    DoHeatmap(gcdata.Tcells, genes.use = genes, slim.col.label = T, group.by = "clonal_group", 
              group.order = unique(clonotypes$clonal_group), cex.row = 15, group.cex = 0, 
              cells.use = cells, title = title, remove.key = T)
    ggsave(paste0(figures.dir, "marker_genes_", title, "_high_vs_low_expansion.png"), width = 6, height = 4, dpi = 200)
  }
}
```


## Compare clonotypes with high cytotoxicity and low cytotoxicity across all individuals
Divide CD4+ and CD8+ clonotypes by whether they have low or high cytotoxicity, and look at differentially expressed genes. Are there differences in KLRB1?
```{r expansion_cytotoxicity}
# Possible DE tests to run in this code chunk.
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]

#  Calculate cytotoxicity score for each T cell.
genes.cytotoxic <- c("PRF1", "GZMA", "GZMK", "NKG7", "GZMH", "CST7")
gcdata.sub <- AddModuleScore(gcdata.sub, genes.list = list(genes.cytotoxic), ctrl.size = 20, enrich.name = "genes_cytotoxic")
gcdata.sub@meta.data <- plyr::rename(gcdata.sub@meta.data, c("genes_cytotoxic1" = "genes_cytotoxic"))

#################
# For all CD4+ T cells excluding Tregs, compare clonotypes with high cytotoxic score to low cytotoxic score.
#################
cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$clonal.CD4 == 1, ])
hist(gcdata.sub@meta.data[cells, "genes_cytotoxic"])
# Identify cells with low and high cytotoxic scores based on cytotoxic score quantile cutoffs.
cutoffs <- quantile(gcdata.sub@meta.data[cells, "genes_cytotoxic"], seq(0, 1, 0.25))
cells.low <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] <= cutoffs[2]]
cells.high <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] > cutoffs[4]]
# cells.low <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] < 0]
# cells.high <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] > 0]
cells <- c(cells.low, cells.high)

# Make a Seurat object with a subset of the data, and with T cells labeled as low or high cytotoxic.
gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
gcdata.Tcells@meta.data[cells.high, "cytotoxic"] <- "high"
gcdata.Tcells@meta.data[cells.low, "cytotoxic"] <- "low"
gcdata.Tcells@meta.data$cytotoxic <- factor(gcdata.Tcells@meta.data$cytotoxic, levels = c("low", "high"))

gcdata.markers <- FindMarkers(gcdata.Tcells, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.Tcells@data>0) > 0.2*length(gcdata.Tcells@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_clonal_CD4_high_vs_low_cytotoxic.csv"))
DoHeatmap(gcdata.Tcells, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, group.by = "cytotoxic", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T)
ggsave(paste0(figures.dir, "top_DE_genes_clonal_CD4_high_vs_low_cytotoxic.png"), width = 4, height = 4, dpi = 200)

#################
# For all CD8+ T cells, compare clonotypes with high cytotoxic score to low cytotoxic score.
#################
cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$clonal.CD8 == 1, ])
hist(gcdata.sub@meta.data[cells, "genes_cytotoxic"])
# Identify cells with low and high cytotoxic scores based on cytotoxic score quantile cutoffs.
cutoffs <- quantile(gcdata.sub@meta.data[cells, "genes_cytotoxic"], seq(0, 1, 0.25))
cells.low <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] <= cutoffs[2]]
cells.high <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] > cutoffs[4]]
# cells.low <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] < 0]
# cells.high <- cells[gcdata.sub@meta.data[cells, "genes_cytotoxic"] > 0]
cells <- c(cells.low, cells.high)

# Make a Seurat object with a subset of the data, and with T cells labeled as low or high cytotoxic.
gcdata.Tcells <- SubsetData(gcdata.sub, cells.use = cells)
gcdata.Tcells@meta.data[cells.high, "cytotoxic"] <- "high"
gcdata.Tcells@meta.data[cells.low, "cytotoxic"] <- "low"
gcdata.Tcells@meta.data$cytotoxic <- factor(gcdata.Tcells@meta.data$cytotoxic, levels = c("low", "high"))

gcdata.markers <- FindMarkers(gcdata.Tcells, ident.1 = cells.high, ident.2 = cells.low, min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.Tcells@data>0) > 0.2*length(gcdata.Tcells@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
write.csv(gcdata.markers, file=paste0(figures.dir, "top_DE_genes_clonal_CD8_high_vs_low_cytotoxic.csv"))
DoHeatmap(gcdata.Tcells, genes.use = rownames(gcdata.markers), slim.col.label = TRUE, group.by = "cytotoxic", cex.row = 15, cex.col = 15, cells.use = cells, remove.key = T)
ggsave(paste0(figures.dir, "top_DE_genes_clonal_CD8_high_vs_low_cytotoxic.png"), width = 6, height = 6, dpi = 200)

#################
# For all CD8+ T cells, look at cytotoxicity score in clonal and non-clonal T cells.
#################
genes.cytotoxic <- c("PRF1", "GZMA", "GZMK", "NKG7", "GZMH", "CST7")
gcdata.sub <- AddModuleScore(gcdata.sub, genes.list = list(genes.cytotoxic), ctrl.size = 20, enrich.name = "genes_cytotoxic")
gcdata.sub@meta.data <- plyr::rename(gcdata.sub@meta.data, c("genes_cytotoxic1" = "genes_cytotoxic"))
VlnPlot(gcdata.sub, features.plot = "genes_cytotoxic", group.by = "clonal.CD8", ident.include = c(1, 2), point.size.use = 0.1)
ggsave(paste0(figures.dir, "vln_cytotoxic_clonal_nonclonal.png"), width = 4, height = 4)
```

## Heat maps of gene expression for cells that are clonotype members.
```{r clonotype_signatures}
genes <- c("CD3D", "CD4", "CD8A", "KLRB1", "GZMA", "GZMB", "PRF1", "IL2", "IFNG", "TNF", "IL10", "TGFB1", "TGFB2", "PDCD1", "CTLA4", "HAVCR2", "LAG3", "TIGIT")
palette = colorRampPalette(c("blue", "white", "red"))(n = 200)  # heatmap color scale

cells <- clonotypes.all$cell_name
cells <- cells[cells %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["CD4_Treg"]], ])]
png(paste0(figures.dir, 'heat_genes_CD4_Treg_clones.png'), width = 6, height = 6, res = 200, units = "in")
heatmap.2(as.matrix(gcdata.sub@scale.data[genes, cells]), trace = "none", dendrogram = "none", col = palette, xlab = "clonotype cell", ylab = "genes", key.xlab = "centered gene expression", labCol = F, margins = c(2, 5))
dev.off()

cells <- clonotypes.all$cell_name
cells <- cells[cells %in% rownames(gcdata@meta.data[gcdata@meta.data$annotate %in% types[["CD8_CD8NK"]], ])]
png(paste0(figures.dir, 'heat_genes_CD8_clones.png'), width = 6, height = 6, res = 200, units = "in")
heatmap.2(as.matrix(gcdata.sub@scale.data[genes, cells]), trace = "none", dendrogram = "none", col = palette, xlab = "clonotype cell", ylab = "genes", key.xlab = "centered gene expression", labCol = F, margins = c(2, 5))
dev.off()
```


## Compare cells with and without TCR reconstruction in cluster of CD8+ T cells with NK markers (CD8+ (P1))
The cells in CD8+ (P1) have NK markers, and only a fraction of them have reconstructed TCRs by TraCeR. Is there any difference between the cells with a reconstructed TCR and the cells without a reconstructed TCR?
```{r CD8P1_TCR}
# Find which CD8+ T cells with NK markers have reconstructed TCR sequences.
cells.CD8P1 <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate == "CD8+ (P1)",])
df.CD8P1 <- data.frame(cell_name = character(), A_productive = character(), B_productive = character())
for (exp in run.exps) {
  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(paste0(proj.path, "/alignreads/tracer/", exp,
                                    "/output/filtered_TCRAB_summary/cell_data.csv"), header = T, sep = ",")
  cells.match <- which(tcr.annotate$cell_name %in% cells.CD8P1)  # find annotation for cells of interest
  df.CD8P1 <- rbind(df.CD8P1, tcr.annotate[cells.match, c("cell_name", "A_productive", "B_productive")])
}
# Keep T cells where at least one of the TCRA or TCRB chains are productively reconstructed.
df.CD8P1 <- df.CD8P1[!(df.CD8P1$A_productive == "") | !(df.CD8P1$B_productive == ""), ]
write.csv(df.CD8P1, file=paste0(figures.dir, "CD8P1_reconstructed_TCRs.csv"), row.names = F, quote = F)

# Do the CD8+ T cells with NK markers that look like NK cells have clonal expansion?
cells.expanded <- clonotypes.all[clonotypes.all$cell_name %in% cells.CD8P1, ]$cell_name
gcdata.sub@meta.data[cells.expanded, c("annotate", "clonal", "clonal.CD4", "clonal.CD8")]
paste0("There are ", length(cells.CD8P1), " CD8+ (P1) cells. ", 
       nrow(df.CD8P1), " of these cells have reconstructed TCR sequences. ", 
       length(cells.expanded), " of these cells are clonally expanded.")

######################################
# Compare CD8+ (P1) cells with a reconstructed TCR and without a reconstructed TCR.
######################################
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]
genes.Tcell <- c("CD3D", "CD3E", "CD3G", "CD8A", "CD8B")
genes.cytotoxic <- c("PRF1", "GZMA", "GZMB", "NKG7")
genes.activateNK <- c("KLRK1", "KLRC2", "KLRC3", "KLRC4", "KLRF1", "SLAMF7")
genes.inhibitNK <- c( "KLRD1", "KLRB1", "KLRC1", "LAIR1")
genes.NKtf <- c("ZBTB16", "IRF8")
genes.Fc <- c("FCER1G", "FCGR3A")

# Make Seurat object of CD8+ (P1) with cells divided into those with and without reconstructed TCR.
cells.TCR <- as.character(df.CD8P1$cell_name)
cells.noTCR <- setdiff(cells.CD8P1, cells.TCR)
cells.use = c(cells.TCR, cells.noTCR)
gcdata.TCR <- SubsetData(gcdata.sub, cells.use = cells.use)
gcdata.TCR@meta.data[cells.TCR, "reconstruct"] = "TCR reconstructed"
gcdata.TCR@meta.data[cells.noTCR, "reconstruct"] = "no TCR reconstructed"
gcdata.TCR <- SetAllIdent(gcdata.TCR, id = "reconstruct")

# Marker genes between cells with and without reconstructed TCR.
p1 <- DoHeatmap(gcdata.TCR, use.scaled = T, genes.use = c(genes.Tcell, genes.cytotoxic, genes.activateNK, genes.inhibitNK, genes.NKtf, genes.Fc), slim.col.label = T, group.by = "reconstruct", title = "CD8+ (P1)")
p1 + scale_fill_gradient2(low = "blue", mid = "white", high = "red") + theme(strip.background = element_blank(), panel.spacing = unit(x = 0.5, units = 'lines')) 
ggsave(paste0(figures.dir, "fig3_CD8P1_with_without_TCR_markergenes.png"), width = 12, height = 8)

# DE genes between cells with and without reconstructed TCR.
gcdata.markers <- FindMarkers(gcdata.TCR, ident.1 = "TCR reconstructed", ident.2 = "no TCR reconstructed", min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.TCR@data>0) > 0.2*length(gcdata.TCR@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
p1 <- DoHeatmap(gcdata.TCR, use.scaled = T, genes.use = rownames(gcdata.markers), slim.col.label = T, group.by = "reconstruct", title = "CD8+ (P1)")
p1 + scale_fill_gradient2(low = "blue", mid = "white", high = "red") + theme(strip.background = element_blank(), panel.spacing = unit(x = 0.5, units = 'lines')) # + scale_fill_distiller(palette = "RdBu")
ggsave(paste0(figures.dir, "fig3_CD8P1_with_without_TCR_DEgenes.png"), width = 12, height = 8)
# write.csv(gcdata.markers, file=paste0(figures.dir, "fig3_top_DE_genes.csv"))

######################################
# Compare CD8+ (P1) cells with high and low CD3 expression. CD3 expression is an indicator that a cell is a T cell.
######################################
df.Tcell <- data.frame("CD3" = Matrix::colMeans(gcdata.sub@data[c("CD3D", "CD3E", "CD3G"), cells.CD8P1]), 
                      "CD8" = Matrix::colMeans(gcdata.sub@data[c("CD8A", "CD8B"), cells.CD8P1]))
ggplot(data = df.Tcell) + geom_point(mapping = aes(x = CD3, y = CD8))
ggsave(paste0(figures.dir, "fig3_CD8P1_CD3_vs_CD8.png"), width = 4, height = 4)
cells.CD3positive <- rownames(df.Tcell[df.Tcell$CD3 > 1, ])  # possibly T cells due to CD3 expression
cells.CD3negative <- rownames(df.Tcell[df.Tcell$CD3 < 1, ])  # unlikely to be T cells
gcdata.TCR <- SubsetData(gcdata.sub, cells.use = c(cells.CD3positive, cells.CD3negative))
gcdata.TCR@meta.data[cells.CD3positive, "CD3"] = "CD3 high"
gcdata.TCR@meta.data[cells.CD3negative, "CD3"] = "CD3 low"
gcdata.TCR <- SetAllIdent(gcdata.TCR, id = "CD3")

gcdata.markers <- FindMarkers(gcdata.TCR, ident.1 = "CD3 high", ident.2 = "CD3 low", min.pct = 0.2, logfc.threshold = 0.5, test.use = test, print.bar = F)
n.genes.test <- length(which(Matrix::rowSums(gcdata.TCR@data>0) > 0.2*length(gcdata.TCR@cell.names)))
gcdata.markers$p_val_adj <- gcdata.markers$p_val * n.genes.test  # Bonferroni correction
gcdata.markers <- gcdata.markers[gcdata.markers$p_val_adj <= 0.1, ]  
gcdata.markers <- gcdata.markers[order(gcdata.markers$avg_logFC, decreasing = T),]
p1 <- DoHeatmap(gcdata.TCR, use.scaled = T, genes.use = rownames(gcdata.markers), slim.col.label = T, group.by = "CD3", title = "CD8+ (P1)")
p1 + scale_fill_gradient2(low = "blue", mid = "white", high = "red") + theme(strip.background = element_blank(), panel.spacing = unit(x = 0.5, units = 'lines')) # + scale_fill_distiller(palette = "RdBu")
ggsave(paste0(figures.dir, "fig3_CD8P1_CD3high_CD3low_DEgenes.png"), width = 12, height = 8)
```

## Examine alpha and beta chains used in reconstructed TCRs of CD8+ T cells
Some of the CD8+ T cells expressing NK markers do have TCRs, suggesting that they are indeed T cells and not NK cells. We find how many CD8+ T cells with NK markers have reconstructed TCRs, and how many of these cells are clonally expanded. Also since we find that these cells have TCRs with different alpha and beta chains, they are unlikely to be MAIT cells. In addition we plot the V and J segment useage within the A and B chains.

```{r CD8_chain}
######################################
# CD8+ cells TCRA TCRB segment usage.
######################################
# load(Rda.subcluster.path)  # cells.CD8 <- rownames(gcdata$CD8@meta.data[gcdata$CD8@meta.data$res.0.6 == 3,])
# Find which CD8+ T cells have reconstructed TCR sequences.
cells.CD8 <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate == "CD8+",])
df.CD8 <- data.frame(cell_name = character(), A_productive = character(), B_productive = character())
for (exp in run.exps) {
  # Since E48 and E60 were later renamed to E48.1 and E48.2, we need to use the original names below 
  # used when initially running tracer, in order to open the tracer files.
  if (exp == "E48.1") {
    exp.filename = "E48"
  } else if (exp == "E48.2") {
    exp.filename = "E60"
  } else {
    exp.filename = exp
  }
  print(paste0(exp, " ", exp.filename))

  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(paste0(proj.path, "/alignreads/tracer/", exp.filename,
                                    "/output/filtered_TCRAB_summary/cell_data.csv"), header = T, sep = ",")
  # Since E48 and E60 were later renamed to E48.1 and E48.2, we need to change the TraCer cell names to match
  # those in the Seurat object gcdata.sub. This will only change names for E48 and E60 to E48.1 and E48.2
  tcr.annotate$cell_name <- sub(paste0("_", exp.filename, "_"), paste0("_", exp, "_"), tcr.annotate$cell_name)

  cells.match <- which(tcr.annotate$cell_name %in% cells.CD8)  # find annotation for cells of interest
  df.CD8 <- rbind(df.CD8, tcr.annotate[cells.match, c("cell_name", "A_productive", "B_productive")])
}
# Keep T cells where at least one of the TCRA or TCRB chains are productively reconstructed.
df.CD8 <- df.CD8[!(df.CD8$A_productive == "") | !(df.CD8$B_productive == ""), ]
paste0("There are ", length(cells.CD8), " CD8+ cells. ", 
       nrow(df.CD8), " of these cells have reconstructed TCR sequences. ")

# Plot the segment useage of V and J segments that are found in TCRA and in TCRB.
# V and J segment useage is visualized with pie charts.
df.A.segments <- data.frame(V = character(), J = character())
df.B.segments <- data.frame(V = character(), J = character())
chains <- as.character(df.CD8$A_productive)
for (chain in chains[chains != ""]) {
  segments <- strsplit(chain, "_")[[1]]
  df.A.segments <- rbind(df.A.segments, data.frame(V = segments[1], J = segments[length(segments)]))
}
chains <- as.character(df.CD8$B_productive)
for (chain in chains[chains != ""]) {
  segments <- strsplit(chain, "_")[[1]]
  df.B.segments <- rbind(df.B.segments, data.frame(V = segments[1], J = segments[length(segments)]))
}
table(df.A.segments$V)
table(df.A.segments$J)
table(df.B.segments$V)
table(df.B.segments$J)

blank_theme <- theme_minimal() + theme(legend.title=element_blank(), axis.title.x = element_blank(), axis.title.y = element_blank(), panel.border = element_blank(), panel.grid=element_blank(), axis.ticks = element_blank(), axis.text.x=element_blank(), plot.title=element_text(size=14, face="bold"))
ggplot(df.A.segments, mapping = aes(x = factor(1), fill = factor(V))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+\nA chain - V segment") + scale_fill_manual(values = material.heat(n_distinct(df.A.segments$V))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8_Achain_Vsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8_Achain_Vsegment.pdf"), width = 6, height = 6)
ggplot(df.A.segments, mapping = aes(x = factor(1), fill = factor(J))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+\nA chain - J segment") + scale_fill_manual(values = material.heat(n_distinct(df.A.segments$J))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8_Achain_Jsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8_Achain_Jsegment.pdf"), width = 6, height = 6)
ggplot(df.B.segments, mapping = aes(x = factor(1), fill = factor(V))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+\nB chain - V segment") + scale_fill_manual(values = material.heat(n_distinct(df.B.segments$V))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8_Bchain_Vsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8_Bchain_Vsegment.pdf"), width = 6, height = 6)
ggplot(df.B.segments, mapping = aes(x = factor(1), fill = factor(J))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+\nB chain - J segment") + scale_fill_manual(values = material.heat(n_distinct(df.B.segments$J))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8_Bchain_Jsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8_Bchain_Jsegment.pdf"), width = 6, height = 6)

######################################
# CD8+ cells with high KLRB1 TCRA TCRB segment usage.
######################################
# Find which CD8+ T cells with high KLRB1 expression have reconstructed TCR sequences.
# Store KLRB1 expression and sort all cells by their KLRB1 expression.
cells.CD8 <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$annotate %in% types[["CD8"]], ])
gcdata.sub@meta.data$KLRB1 <- gcdata.sub@data["KLRB1", ]
df.sort <- gcdata.sub@meta.data[order(gcdata.sub@meta.data[, "KLRB1"]),]  
df.plot <- data.frame("KLRB1" = gcdata.sub@data["KLRB1", cells.CD8])
ggplot(data = df.plot) + geom_histogram(mapping = aes(x = KLRB1), binwidth = 0.25, color="black", fill="white") + 
  labs(x = "KLRB1 expression in CD8+ cells", y = "count") 
ggsave(paste0(figures.dir, "KLRB1_CD8_expression_histogram.png"), width = 3, height = 3)

# Identify cells with high KLRB1 based on absolute cutoffs.
cells.sort <- rownames(df.sort[df.sort$annotate %in% types[["CD8"]], ])  # subset on CD8+ T cells
cells.CD8 <- cells.sort[df.sort[cells.sort, "KLRB1"] > 2]

# Collect TCR segments for CD8+ cells with high KLRB1.
df.CD8 <- data.frame(cell_name = character(), A_productive = character(), B_productive = character())
for (exp in run.exps) {
  # Since E48 and E60 were later renamed to E48.1 and E48.2, we need to use the original names below 
  # used when initially running tracer, in order to open the tracer files.
  if (exp == "E48.1") {
    exp.filename = "E48"
  } else if (exp == "E48.2") {
    exp.filename = "E60"
  } else {
    exp.filename = exp
  }
  print(paste0(exp, " ", exp.filename))

  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(paste0(proj.path, "/alignreads/tracer/", exp.filename,
                                    "/output/filtered_TCRAB_summary/cell_data.csv"), header = T, sep = ",")
  # Since E48 and E60 were later renamed to E48.1 and E48.2, we need to change the TraCer cell names to match
  # those in the Seurat object gcdata.sub. This will only change names for E48 and E60 to E48.1 and E48.2
  tcr.annotate$cell_name <- sub(paste0("_", exp.filename, "_"), paste0("_", exp, "_"), tcr.annotate$cell_name)

  cells.match <- which(tcr.annotate$cell_name %in% cells.CD8)  # find annotation for cells of interest
  df.CD8 <- rbind(df.CD8, tcr.annotate[cells.match, c("cell_name", "A_productive", "B_productive")])
}
# Keep T cells where at least one of the TCRA or TCRB chains are productively reconstructed.
df.CD8 <- df.CD8[!(df.CD8$A_productive == "") | !(df.CD8$B_productive == ""), ]
paste0("There are ", length(cells.CD8), " CD8+ cells with high KLRB1. ", 
       nrow(df.CD8), " of these cells have reconstructed TCR sequences. ")

# Plot the segment useage of V and J segments that are found in TCRA and in TCRB.
# V and J segment useage is visualized with pie charts.
df.A.segments <- data.frame(V = character(), J = character())
df.B.segments <- data.frame(V = character(), J = character())
chains <- as.character(df.CD8$A_productive)
for (chain in chains[chains != ""]) {
  segments <- strsplit(chain, "_")[[1]]
  df.A.segments <- rbind(df.A.segments, data.frame(V = segments[1], J = segments[length(segments)]))
}
chains <- as.character(df.CD8$B_productive)
for (chain in chains[chains != ""]) {
  segments <- strsplit(chain, "_")[[1]]
  df.B.segments <- rbind(df.B.segments, data.frame(V = segments[1], J = segments[length(segments)]))
}
table(df.A.segments$V)
table(df.A.segments$J)
table(df.B.segments$V)
table(df.B.segments$J)

blank_theme <- theme_minimal() + theme(legend.title=element_blank(), axis.title.x = element_blank(), axis.title.y = element_blank(), panel.border = element_blank(), panel.grid=element_blank(), axis.ticks = element_blank(), axis.text.x=element_blank(), plot.title=element_text(size=14, face="bold"))
ggplot(df.A.segments, mapping = aes(x = factor(1), fill = factor(V))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+ with high KLRB1\nA chain - V segment") + scale_fill_manual(values = material.heat(n_distinct(df.A.segments$V))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Achain_Vsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Achain_Vsegment.pdf"), width = 6, height = 6)
ggplot(df.A.segments, mapping = aes(x = factor(1), fill = factor(J))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+ with high KLRB1\nA chain - J segment") + scale_fill_manual(values = material.heat(n_distinct(df.A.segments$J))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Achain_Jsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Achain_Jsegment.pdf"), width = 6, height = 6)
ggplot(df.B.segments, mapping = aes(x = factor(1), fill = factor(V))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+ with high KLRB1\nB chain - V segment") + scale_fill_manual(values = material.heat(n_distinct(df.B.segments$V))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Bchain_Vsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Bchain_Vsegment.pdf"), width = 6, height = 6)
ggplot(df.B.segments, mapping = aes(x = factor(1), fill = factor(J))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD8+ with high KLRB1\nB chain - J segment") + scale_fill_manual(values = material.heat(n_distinct(df.B.segments$J))) + blank_theme
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Bchain_Jsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig3_CD8highKLRB1_Bchain_Jsegment.pdf"), width = 6, height = 6)
```

## Examine alpha and beta chains used in reconstructed TCRs of CD4+ T cells
We want to demonstrate that the CD4+ T cells with high KLRB1 are not MAIT cells, so we will look at their reconstructed TCRA and TCRB chains. In addition we plot the V and J segment useage within the A and B chains.

```{r CD4KLRB1_chain}
# Find which CD4+ T cells have reconstructed TCR sequences. Cluster 3 has high KLRB1.
cells.CD4 <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$res.1 == 3,])
df.CD4 <- data.frame(cell_name = character(), A_productive = character(), B_productive = character())
for (exp in run.exps) {
  # Since E48 and E60 were later renamed to E48.1 and E48.2, we need to use the original names below 
  # used when initially running tracer, in order to open the tracer files.
  if (exp == "E48.1") {
    exp.filename = "E48"
  } else if (exp == "E48.2") {
    exp.filename = "E60"
  } else {
    exp.filename = exp
  }
  print(paste0(exp, " ", exp.filename))

  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(paste0(proj.path, "/alignreads/tracer/", exp.filename,
                                    "/output/filtered_TCRAB_summary/cell_data.csv"), header = T, sep = ",")
  # Since E48 and E60 were later renamed to E48.1 and E48.2, we need to change the TraCer cell names to match
  # those in the Seurat object gcdata.sub. This will only change names for E48 and E60 to E48.1 and E48.2
  tcr.annotate$cell_name <- sub(paste0("_", exp.filename, "_"), paste0("_", exp, "_"), tcr.annotate$cell_name)

  cells.match <- which(tcr.annotate$cell_name %in% cells.CD4)  # find annotation for cells of interest
  df.CD4 <- rbind(df.CD4, tcr.annotate[cells.match, c("cell_name", "A_productive", "B_productive")])
}
# Keep T cells where at least one of the TCRA or TCRB chains are productively reconstructed.
df.CD4 <- df.CD4[!(df.CD4$A_productive == "") | !(df.CD4$B_productive == ""), ]
write.csv(df.CD4, file=paste0(figures.dir, "CD4KLRB1_reconstructed_TCRs.csv"), row.names = F, quote = F)

# Do the CD4+ T cells with high KLRB1 expression have clonal expansion?
cells.expanded <- clonotypes.all[clonotypes.all$cell_name %in% cells.CD4, ]$cell_name
gcdata.sub@meta.data[cells.expanded, c("annotate", "clonal", "clonal.CD4", "clonal.CD8")]

paste0("There are ", length(cells.CD4), " CD4+ T cells with high KLRB1. ", 
       nrow(df.CD4), " of these cells have reconstructed TCR sequences. ", 
       length(cells.expanded), " of these cells are clonally expanded.")

# Plot the segment useage of V and J segments that are found in TCRA and in TCRB.
# V and J segment useage is visualized with pie charts.
df.A.segments <- data.frame(V = character(), J = character())
df.B.segments <- data.frame(V = character(), J = character())
chains <- as.character(df.CD4$A_productive)
for (chain in chains[chains != ""]) {
  segments <- strsplit(chain, "_")[[1]]
  df.A.segments <- rbind(df.A.segments, data.frame(V = segments[1], J = segments[length(segments)]))
}
chains <- as.character(df.CD4$B_productive)
for (chain in chains[chains != ""]) {
  segments <- strsplit(chain, "_")[[1]]
  df.B.segments <- rbind(df.B.segments, data.frame(V = segments[1], J = segments[length(segments)]))
}
table(df.A.segments$V)
table(df.A.segments$J)
table(df.B.segments$V)
table(df.B.segments$J)

blank_theme <- theme_minimal() + theme(legend.title=element_blank(), axis.title.x = element_blank(), axis.title.y = element_blank(), panel.border = element_blank(), panel.grid=element_blank(), axis.ticks = element_blank(), axis.text.x=element_blank(), plot.title=element_text(size=14, face="bold"))
ggplot(df.A.segments, mapping = aes(x = factor(1), fill = factor(V))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD4+ T cells with high KLRB1\nA chain - V segment") + scale_fill_manual(values = material.heat(n_distinct(df.A.segments$V))) + blank_theme
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Achain_Vsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Achain_Vsegment.pdf"), width = 6, height = 6)
ggplot(df.A.segments, mapping = aes(x = factor(1), fill = factor(J))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD4+ T cells with high KLRB1\nA chain - J segment") + scale_fill_manual(values = material.heat(n_distinct(df.A.segments$J))) + blank_theme
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Achain_Jsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Achain_Jsegment.pdf"), width = 6, height = 6)
ggplot(df.B.segments, mapping = aes(x = factor(1), fill = factor(V))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD4+ T cells with high KLRB1\nB chain - V segment") + scale_fill_manual(values = material.heat(n_distinct(df.B.segments$V))) + blank_theme
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Bchain_Vsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Bchain_Vsegment.pdf"), width = 6, height = 6)
ggplot(df.B.segments, mapping = aes(x = factor(1), fill = factor(J))) + geom_bar(width = 1) + coord_polar(theta = "y") + ggtitle("CD4+ T cells with high KLRB1\nB chain - J segment") + scale_fill_manual(values = material.heat(n_distinct(df.B.segments$J))) + blank_theme
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Bchain_Jsegment.png"), width = 6, height = 6)
ggsave(paste0(figures.dir, "fig4_CD4KLRB1_Bchain_Jsegment.pdf"), width = 6, height = 6)
```

## Clustering analysis of clonally expanded T cells
Take the subset of T cells that are clonally expanded and cluster them to identify transcriptional profiles. There may be profiles we identify that were not as obvious when clonal and non clonally expanded cells were clustered all together.
```{r cluster_clonal}
prefix <- "clonalTcells_"  # use in figure file names
# gcdata <- SetAllIdent(gcdata, id = "res.1")

# Make Seurat object consisting only of clonally expanded T cells.
# Cells and genes have been filtered for quality and 
# the expression data is already normalized and log transformed.
cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$clonal == 1, ])
gcdata <- SubsetData(gcdata.sub, cells.use = cells)

# Variable gene selection approach from Chris Smillie where we find variable genes common across experiments.
# Only use experiments with at least 10 clonal T cells when defining variable genes common across experiments.
exps.use <- names(which(table(gcdata@meta.data$experiment) >= 10)) 
gcdata.var <- SubsetData(gcdata, cells.use = rownames(gcdata@meta.data[gcdata@meta.data$experiment %in% exps.use, ]))
gcdata@var.genes <- get_var_genes(gcdata.var, gcdata.var@meta.data$experiment, method="seurat", do.plot=T,
                                  num_genes=1500, prefix=figures.path, do.flatten=T)
rm(gcdata.var)

# Number of variable genes
print(length(gcdata@var.genes))

# Batch correction scaling.
# Initialize empty object@scale.data matrix.
gcdata@scale.data <- matrix(0, nrow(gcdata@data), ncol(gcdata@data))
rownames(gcdata@scale.data) <- rownames(gcdata@data)
colnames(gcdata@scale.data) <- colnames(gcdata@data)

# Within each batch, center the gene expression by subtracting from a gene its mean 
# gene expression across all cells in the batch.
for (ind in unique(experiment)) {
  cells <- which(gcdata@meta.data$experiment == ind)
  data.ind <- as.matrix(gcdata@data[, cells])
  gcdata@scale.data[, cells] <- t(scale(t(data.ind), center = T, scale = F))
  print(ind)
  print(dim(gcdata@scale.data[,cells]))
}
range(rowSums(gcdata@scale.data[,gcdata@meta.data$experiment == "E92"]))

# Do PCA using variable genes or using all genes (pc.genes = rownames(object@data))
gcdata <- RunPCA(gcdata, pc.genes = gcdata@var.genes, pcs.compute = 30, do.print = TRUE, pcs.print = 5, genes.print = 5)

# For each gene (variable and non-variable), score correlation between its
# expression across cells and its corresponding PCi score. This is done for all
# PCi (i from 1 to n), so the final matrix dimension is # genes x # PCs.
gcdata <- ProjectPCA(gcdata, genes.print = 5)

# Jackstraw analysis
# gcdata <- JackStraw(gcdata, num.replicate = 10, num.pc = 30)

# PCA plots
VizPCA(gcdata, pcs.use = 1:2)
print(PCAPlot(gcdata, dim.1 = 1, dim.2 = 2, pt.size = 1, use.full = FALSE, do.return = TRUE))
ggsave(paste0(figures.dir, prefix, "PCA.png"), width = 6, height = 6)

# Identify significant principal components
PrintPCA(gcdata, pcs.print = 1:5, genes.print = 10, use.full = FALSE)

png(paste0(figures.dir, prefix, "heat.png"), width = 1800, height = 1800, res = 300)
PCHeatmap(gcdata, pc.use = 1:6, cells.use = 100, do.balanced = TRUE)
dev.off()

print(PCElbowPlot(gcdata))
ggsave(paste0(figures.dir, prefix, "screeplot.png"), width = 6, height = 6)

# print(JackStrawPlot(gcdata, PCs = 1:12))
# ggsave(paste0(figures.dir, prefix, "jackstraw.png"), width = 6, height = 6)

# PC correlation with number of reads and number of expressed genes
gcdata <- SetIdent(gcdata, ident.use = 1)
counts <- as.data.frame(as.matrix(gcdata@data))
PCPlots(counts, gcdata, no.pcs=6, plotType='all', outDirectory=figures.dir, fileType = '.png', colors = c("blue"))

gcdata <- FindClusters(gcdata, reduction.type = "pca", dims.use = 1:20, resolution = 1, print.output = F, save.SNN = T, force.recalc = T)
PrintFindClustersParams(gcdata)

gcdata.markers <- FindAllMarkers(gcdata, min.pct = 0.1, logfc.threshold = 0.5, test.use = test, 
                                 return.thresh = 0.1/nrow(gcdata@data), print.bar = T)
top2genes <- gcdata.markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(2, avg_logFC)
top15genes <- gcdata.markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(15, avg_logFC)
top100genes <- gcdata.markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(100, avg_logFC)

# Write cluster markers to csv file
write.csv(top100genes, file=paste0(figures.dir, prefix, "top_DE_genes.csv"))

# tSNE visualization
gcdata <- RunTSNE(gcdata, dims.use = 1:20, do.fast = T, reduction.use = "pca", perplexity = 30)

# Save current progress Rda.path
save(gcdata, file = Rda.clonalTcells.path)

# Color tSNE plot by cluster or by experiment (there are multiple samples per experiment)
df <- data.frame(gcdata@dr$tsne@cell.embeddings, ident = gcdata@ident, experiment  = gcdata@meta.data$experiment)
n.clusters <- length(unique(df$ident))
n.replicates <- length(unique(df$experiment))
centers <- df %>% group_by(ident) %>% dplyr::summarise(x = median(tSNE_1), y = median(tSNE_2))  # cluster labels
p1 <- ggplot(data = df) + geom_point(mapping = aes(tSNE_1, tSNE_2, color = ident), size = 2) + 
  labs(color = "cluster") + scale_colour_manual(values = material.heat(n.clusters)) + 
  geom_text(data = centers, mapping = aes(x = x, y = y, label = ident), size = 8) +
  guides(colour = guide_legend(override.aes = list(size=5)))
p2 <- ggplot(data = df) + geom_point(mapping = aes(tSNE_1, tSNE_2, color = experiment), size = 2, alpha = 0.75) + guides(col = guide_legend(ncol = 1)) + scale_colour_manual(values = material.heat(n.replicates))
tsne <- plot_grid(p1, p2, align = 'h', labels = c('A', 'B'))
#print(tsne)
save_plot(paste0(figures.dir, prefix, 'tSNE_cluster_experiment.png'), tsne, base_aspect_ratio = 2, base_height = 10)

# G1/S/G2 phase.
gcdata <- CellCycleScoring(object = gcdata, s.genes = s.genes, g2m.genes = g2m.genes, set.ident = T)
FeaturePlot(gcdata, features.plot = "G2M.Score", pt.size = 2)
FeaturePlot(gcdata, features.plot = "S.Score", pt.size = 2)
TSNEPlot(gcdata, pt.size = 1, colors.use = material.heat(3))
ggsave(paste0(figures.dir, prefix, "tSNE_cell_cycle.png"), width = 4, height = 4, dpi = 200)
gcdata <- SetAllIdent(gcdata, id = "old.ident")

# Marker gene expression heat map.
p <- DoHeatmap(gcdata, genes.use = top15genes$gene, slim.col.label = TRUE, cex.row = 15, remove.key = TRUE)
save_plot(paste0(figures.dir, prefix, "heat_expression.png"), p, base_aspect_ratio = .9, base_height = 15)

VlnPlot(gcdata, c("CD4", "CD8A"), nCol = 1, 
          cols.use = material.heat(n_distinct(gcdata@ident)))
ggsave(paste0(figures.dir, prefix, "vln_CD4_CD8.png"), width = 6, height = 8)

MultipleFeaturePlot(gcdata, c("CD6"), feature.type = "gene", pt.size = 2)
ggsave(paste0(figures.dir, prefix, "tSNE_CD6.png"), width = 4, height = 4)
MultipleFeaturePlot(gcdata, c("CCL4", "CCL5", "CD4", "CD8A"), feature.type = "gene", pt.size = 2)
ggsave(paste0(figures.dir, prefix, "tSNE_CCL4_CCL5.png"), width = 8, height = 8)

# Cytotoxicity of clonally expanded T cells.
genes.cytotoxic <- c("PRF1", "GZMA", "GZMK", "NKG7", "GZMH", "CST7")
gcdata@meta.data$genes_cytotoxic <- NULL
gcdata <- AddModuleScore(gcdata, genes.list = list(genes.cytotoxic), ctrl.size = 20, enrich.name = "genes_cytotoxic")
gcdata@meta.data <- plyr::rename(gcdata@meta.data, c("genes_cytotoxic1" = "genes_cytotoxic"))
MultipleFeaturePlot(gcdata, c("genes_cytotoxic"), feature.type = "meta", pt.size = 2)
ggsave(paste0(figures.dir, prefix, "_tSNE_cytotoxic_score.png"), width = 4, height = 4)
MultipleFeaturePlot(gcdata, c("CD4", "CD8A"), feature.type = "gene", pt.size = 2)
ggsave(paste0(figures.dir, prefix, "tSNE_CD4_CD8A.png"), width = 8, height = 4)

```

## Clustering analysis of clonally expanded T cells using non-negative matrix factorization
Cluster clonally expanded T cells using NMF with a range of ranks. The rank is the number of metagenes and the number of clusters. The choice of the rank is heuristic. For each clustering, we also calculate DE genes.
```{r cluster_clonal_NMF, eval = T}
prefix <- "clonalTcells_"  # use in figure file names
# gcdata <- SetAllIdent(gcdata, id = "NMF_rank4")
# gcdata <- SetAllIdent(gcdata, id = "res.1")

# Make Seurat object consisting only of clonally expanded T cells.
# Cells and genes have been filtered for quality and 
# the expression data is already normalized and log transformed.
cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$clonal == 1, ])
gcdata <- SubsetData(gcdata.sub, cells.use = cells)

# Vector of ranks for which to calculate NMF.
rank <- 3:4

# Possible DE tests to run in this code chunk.
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]

# Perform 10 NMF runs for each value of rank r.
x <- as.matrix(gcdata@data[gcdata@var.genes, ])
x <- x[which(rowSums(x) > 0), ]
estim.r <- nmf(x, rank = rank, method = "brunet", nrun = 10, seed = 123456, .options = "tvP")
png(paste0(figures.dir, "NMF_plot_rank.png"), width = 2000, height = 1000, res = 100)
plot(estim.r)
dev.off()
# png(paste0(figures.dir, "NMF_consensusmap_rank.png"), width = 2000, height = 1000, res = 100)
# consensusmap(estim.r)
# dev.off()

# Iterate over different ranks and look at clustering.
for (i in seq(rank)) {
  r <- rank[i]
  
  # Plot coefficient values of metagenes for each cell.
  png(paste0(figures.dir, "NMF_coefmap_rank", r, ".png"), width = 2000, height = 1000, res = 100)
  coefmap(estim.r$fit[[i]], Colv=T)
  dev.off()

  # Show genes contributing the most to each metagene (factor). 
  # Remember, this is limited to the variable genes if that was the input to NMF.
  for (j in seq(r)) {
    top.features <- names(sort(estim.r$fit[[i]]@fit@W[, j], decreasing = T)[1:10])
    print(paste(top.features, collapse = " "))
  }
  s <- extractFeatures(estim.r$fit[[i]], 10L)
  for (j in seq(r)) {
    top.features <- rownames(estim.r$fit[[i]]@fit@W)[s[[j]]]
    print(paste(top.features, collapse = " "))
  }
  
  # Plot NMF residuals.
  png(paste0(figures.dir, "NMF_residuals_rank", r, ".png"), width = 2000, height = 1000, res = 100)
  plot(estim.r$fit[[i]])
  dev.off()

  # Check coefficient matrix has cell names in same order as Seurat object gcdata.
  H <- estim.r$fit[[i]]@fit@H
  all.equal(as.integer(as.character(predict(estim.r$fit[[i]], what = "samples"))), as.integer(apply(H, 2, which.max)))
  all.equal(colnames(H), colnames(gcdata@data))
  all.equal(rownames(estim.r$fit[[i]]@fit@W), gcdata@var.genes)
  
  # Cluster cells based on maximum coefficient for metagene.
  gcdata@meta.data[[paste0("NMF_rank", r)]] <- predict(estim.r$fit[[i]], what = "samples")
  gcdata <- SetAllIdent(gcdata, paste0("NMF_rank", r))

  # tSNE visualization and CD4 and CD8 expression across clusters and find DE genes.
  TSNEPlot(gcdata, colors.use = material.heat(r))
  ggsave(paste0(figures.dir, "tSNE_NMF_rank", r, ".png"), width = 6, height = 6)
  VlnPlot(gcdata, c("CD4", "CD8A"), nCol = 1, 
          cols.use = material.heat(n_distinct(gcdata@ident)))
  ggsave(paste0(figures.dir, "vln_NMF_rank", r, ".png"), width = 8, height = 8)
  
  markers <- FindAllMarkers(gcdata, min.pct = 0.2, logfc.threshold = 0.5, only.pos = T, test.use = test, return.thresh = 0.1/nrow(gcdata@data), print.bar = T)
  top15genes <- markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(15, avg_logFC)
  top100genes <- markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(100, avg_logFC)
  write.csv(markers, file=paste0(figures.dir, "top_DE_genes_NMF_rank", r, ".csv"))
  p <- DoHeatmap(gcdata, genes.use = top15genes$gene, slim.col.label = TRUE, remove.key = TRUE, cex.row = 15, cex.col = 15) + scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
  save_plot(paste0(figures.dir, "heat_expression_NMF_rank", r, ".png"), p, base_aspect_ratio = .9, base_height = 15)

  # Adjusted rand index test for overlap between NMF and Louvain cluster labelings. 
  # This goes between 0 (completely dissimilar clustering) to 1 (identical clustering). 
  # The adjustment corrects for chance grouping between cluster elements.
  # ari <- dplyr::select(gcdata@meta.data, paste0("NMF_rank", r), res.1)
  # adj.rand.index(as.numeric(ari[[paste0("NMF_rank", r)]]), as.numeric(ari$res.1))
}

# Save results including NMF and Seurat object with clustering using different ranks.
gcdata <- SetAllIdent(gcdata, "res.1")
save(gcdata, file = Rda.clonalTcells.path)  # identity in NMF_rankr
save(estim.r, file = Rda.clonalTcells.NMF.path)
```

## Highlight marker genes in clonally expanded T cells (NMF clustering)
Annotate clonally expanded T cells with NMF rank 4 and plot tSNE. Look at marker genes selected by Kai in only the clonally expanded T cells.
```{r marker_clonally_expanded}
load(Rda.clonalTcells.path)  # identity in NMF_rankr
gcdata <- SetAllIdent(gcdata, id = "NMF_rank4")

# Possible DE tests to run in this code chunk.
tests <- c("wilcox", "t", "roc", "MAST", "bimod")
test <- tests[2]

# Cluster 3 is mostly cycling T cells.
# G1/S/G2 phase
gcdata <- CellCycleScoring(object = gcdata, s.genes = s.genes, g2m.genes = g2m.genes, set.ident = T)
FeaturePlot(gcdata, features.plot = "G2M.Score", pt.size = 2)
FeaturePlot(gcdata, features.plot = "S.Score", pt.size = 2)
TSNEPlot(gcdata, pt.size = 1)
ggsave(paste0(figures.dir, "tSNE_NMF_clonally_expanded_cell_cycle.png"), width = 4, height = 4, dpi = 200)
gcdata <- SetAllIdent(gcdata, id = "NMF_rank4")
freq.phase.cluster <- table(gcdata@meta.data$Phase, gcdata@meta.data$NMF_rank4)
freq.phase.cluster <- sweep(freq.phase.cluster, 2, colSums(freq.phase.cluster), '/') 
colSums(freq.phase.cluster[2:3,])  # cells in G2M or S phase.

# DE genes.
current.cluster.ids <- c(1, 2, 3, 4)  
new.cluster.ids <- c(3, 1, 4, 2)
gcdata@ident <- plyr::mapvalues(x = gcdata@ident, from = current.cluster.ids, to = new.cluster.ids)
gcdata@ident <- factor(gcdata@ident, levels = sort(levels(gcdata@ident)))
gcdata <- StashIdent(gcdata, save.name = "NMF_rank4")
markers <- FindAllMarkers(gcdata, min.pct = 0.2, logfc.threshold = 0.5, only.pos = T, test.use = test, return.thresh = 0.1/nrow(gcdata@data), print.bar = T)
n.genes.test <- length(which(Matrix::rowSums(gcdata@data>0) > 0.2*length(gcdata@cell.names)))
markers$p_val_adj <- markers$p_val * n.genes.test  # Bonferroni correction
markers <- markers[markers$p_val_adj <= 0.1, ]  
top15genes <- markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(15, avg_logFC)
top100genes <- markers %>% filter(avg_logFC > 0) %>% group_by(cluster) %>% top_n(100, avg_logFC)
write.csv(markers, file=paste0(figures.dir, "fig2_top_DE_genes_clonalTcells_NMF_rank4.csv"))

# gcdata.markers.write <- gcdata.markers %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(-avg_logFC, .by_group = T)

# Annotate cell subsets.
current.cluster.ids <- c(1, 2, 3, 4)
new.cluster.ids <- c("CD8+", "CD4+", "Treg", "Cycling T cells")
gcdata@ident <- plyr::mapvalues(x = gcdata@ident, from = current.cluster.ids, to = new.cluster.ids)

# Plot cells with annotation.
df <- data.frame(gcdata@dr$tsne@cell.embeddings, ident = gcdata@ident)
n.clusters <- length(unique(df$ident))
df$ident <- factor(df$ident, levels = c("CD8+", "CD4+", "Treg", "Cycling T cells"))
centers <- df %>% group_by(ident) %>% dplyr::summarise(x = median(tSNE_1), y = median(tSNE_2))  # cluster labels
p1 <- ggplot(data = df) + geom_point(mapping = aes(tSNE_1, tSNE_2, color = ident), size = 2) + 
  labs(color = "cluster") + scale_colour_manual(values = material.heat(n.clusters)) + 
  theme(legend.text=element_text(size=20))
ggsave(paste0(figures.dir, 'fig2_tSNE_heat_clonally_expanded_annotate.png'), width = 6, height = 6)
ggsave(paste0(figures.dir, 'fig2_tSNE_heat_clonally_expanded_annotate.pdf'), width = 6, height = 6)
print(p1)

# Plot selected marker genes. 
genes <- c("KLRK1", "KLRD1", "CCL5", "CCL4", "FASLG", "KLRB1", "PTGER2", "IL18RAP", "CCR6", "CD40LG", "IL7R", "ID2", "FOXP3", "TNFRSF18", "IL2RA", "CTLA4", "CCR8", "MCM7", "PCNA", "TYMS")
DoHeatmap(gcdata, genes.use = genes, slim.col.label = T, group.label.rot = T, cex.row = 15, group.cex = 20, 
          group.order = c("CD8+", "CD4+", "Treg", "Cycling T cells"), remove.key = T, group.spacing = 0.75) + 
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
ggsave(paste0(figures.dir, "fig2_heat_clonally_expanded_markers.png"), width = 8, height = 8)
ggsave(paste0(figures.dir, "fig2_heat_clonally_expanded_markers.pdf"), width = 8, height = 8)

# Plot mean of selected marker genes. 
gcdata.plot <- AverageExpression(gcdata, genes.use = genes, return.seurat = T)
DoHeatmap(gcdata.plot, genes.use = genes, use.scaled = T, slim.col.label = T, group.label.rot = T, cex.row = 15, group.cex = 20, remove.key = F) + scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) + theme(strip.background = element_blank(), panel.spacing = unit(x = 0.5, units = 'lines'))
ggsave(paste0(figures.dir, "fig2_heat_mean_clonally_expanded_markers.png"), width = 4, height = 12)
ggsave(paste0(figures.dir, "fig2_heat_mean_clonally_expanded_markers.pdf"), width = 4, height = 12)

# Plot top DE genes.
p <- DoHeatmap(gcdata, genes.use = top15genes$gene, slim.col.label = TRUE, remove.key = TRUE, group.label.rot = T, cex.row = 15, group.cex = 20, group.order = c("CD8+", "CD4+", "Treg", "Cycling T cells")) + 
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0)
save_plot(paste0(figures.dir, "fig2_heat_expression_NMF_rank4.png"), p, base_aspect_ratio = .9, base_height = 15)
save_plot(paste0(figures.dir, "fig2_heat_expression_NMF_rank4.pdf"), p, base_aspect_ratio = .9, base_height = 15)

gcdata <- SetAllIdent(gcdata, id = "NMF_rank4")
```

## OLD DO NOT USE TraCeR compare clonal and non-clonal expansions: all experiments
For each gene, look at mean gene expression and variability of that gene in clonally expanded cells compared to non-expanded cells.

```{r tracer_clonal_allexps, eval = F}
set.seed(100)

# Make Seurat object with only data from chosen experiments.
run.exps <- unique(experiment)
cells.exps <- rownames(gcdata@meta.data[gcdata@meta.data$experiment %in% run.exps, ])
gcdata.sub <- SubsetData(gcdata, cells.use = cells.exps)

# Expression matrix
gcdata.sub.matrix <- as.matrix(gcdata.sub@data)
#gcdata.sub.matrix <- as.matrix(gcdata.sub@data[gcdata.sub@var.genes, ])  # only look at variable genes

# Store mean and variance of expression for clonal and nonclonal cells for all experiments.
mean.clonal <- matrix(, nrow = nrow(gcdata.sub.matrix), ncol = 0)
mean.nonclonal <- matrix(, nrow = nrow(gcdata.sub.matrix), ncol = 0)
var.clonal <- matrix(, nrow = nrow(gcdata.sub.matrix), ncol = 0)
var.nonclonal <- matrix(, nrow = nrow(gcdata.sub.matrix), ncol = 0)

for (exp in run.exps) {
  print(exp)
  
  # Find names of clonotypes that are above minimum group_size cutoff.
  group_size.all <- sapply(clonotypes.cells[[exp]], function(clone) clone$group_size[1])
  clonotypes.minsize <- names(which(group_size.all >= group_size.cutoff))

  # Skip rest of analysis for this experiment if no clonotypes above minimum group_size cutoff were found.
  if (!length(clonotypes.minsize)) {next}

  # Identify in which clonotypes the majority of cells belong to the requested T cell types.
  indicator <- sapply(clonotypes.cells[[exp]][clonotypes.minsize], function(clone) { 
    group.size <- clone$group_size[1]  # number of cells in clonotype
    cells.annotate <- gcdata.sub@meta.data[as.vector(clone$cell_name), ]$annotate  # annotated identity of cells
    num.cell.type <- sum(cells.annotate %in% cell.type)  # number of cells with requested T cell type
    ifelse(num.cell.type/group.size >= 0.5, T, F) 
    })
  clonotypes.request <- names(which(indicator))
  
  # Skip rest of analysis for this experiment if no clonotypes from requested T cell type were found.
  if (!length(clonotypes.request)) {next}
  
  # Get mean gene expression and variance of gene expression within each clonotype. 
  # Row is gene and column is mean or variance of expression within each clonotype.
  cells.clonal <- lapply(clonotypes.request, function(clone) 
    as.vector(clonotypes.cells[[exp]][[clone]]$cell_name))
  names(cells.clonal) <- clonotypes.request  #THERE ARE ISSUES WHEN THE NUMBER OF CLONOTYPES IS 1. CHECK E83
  exp.mean.clonal <- sapply(cells.clonal, function(cells) apply(gcdata.sub.matrix[, cells], 1, mean))
  exp.var.clonal <- sapply(cells.clonal, function(cells) apply(gcdata.sub.matrix[, cells], 1, var))
  colnames(exp.mean.clonal) <- paste0(exp, "_", colnames(exp.mean.clonal))
  colnames(exp.var.clonal) <- paste0(exp, "_", colnames(exp.var.clonal))

  ##################
  # Get mean and variance of gene expression from nonclonal cells.
  ##################
  # Find all nonclonal cells in this experiment.
  cells.exp <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$experiment == exp, ])  # all cells in exp
  cells.clonal <- unlist(sapply(clonotypes.cells[[exp]], function(clone) as.vector(clone$cell_name)))
  cells.nonclonal <- setdiff(cells.exp, cells.clonal)

  # Get size of clonotypes that are above minimum group_size cutoff and have T cell type to analyze.
  # This will be used in drawing random sets of nonclonal size as the same size as the clonotypes.
  group_size <- sapply(clonotypes.request, function(clone) clonotypes.cells[[exp]][[clone]]$group_size[1])
  
  # Initialize matrices to store the mean and variance of expression across several random runs
  # of choosing random sets of nonclonal cells.
  # Dimensions are ngenes x nclonotypes.request, and values are initialized to 0.  
  exp.mean.nonclonal <- matrix(0, nrow(exp.mean.clonal), length(group_size))
  exp.var.nonclonal <- matrix(0, nrow(exp.mean.clonal), length(group_size))

  # Choose a random set of nonclonal cells, with the number of cells matching the number of cells
  # in a clonotype. Do this multiple for multiple runs.
  for (i in seq(n.runs)) {
    cells.nonclonal.random <- lapply(group_size, function(size) sample(cells.nonclonal, size, replace = F))
    names(cells.nonclonal.random) <- clonotypes.request  # CHECK E83 works
    exp.mean.rand <- sapply(cells.nonclonal.random, function(cells) 
      apply(gcdata.sub.matrix[, cells], 1, mean))
    exp.var.rand <- sapply(cells.nonclonal.random, function(cells) 
      apply(gcdata.sub.matrix[, cells], 1, var))
    colnames(exp.mean.rand) <- paste0(exp, "_", colnames(exp.mean.rand))
    colnames(exp.var.rand) <- paste0(exp, "_", colnames(exp.var.rand))

    # Add the mean and variance to existing mean and variance matrix, which were initialized to 0.
    # After the loop, take an average by dividing by the number of runs.        
    exp.mean.nonclonal <- exp.mean.nonclonal + exp.mean.rand
    exp.var.nonclonal <- exp.var.nonclonal + exp.var.rand
  }
  # Take average across the iterations of choosing random sets of nonclonal cells.
  exp.mean.nonclonal <- exp.mean.nonclonal / n.runs
  exp.var.nonclonal <- exp.var.nonclonal / n.runs
  
  # Add mean and variance of expression to matrix containing data for all experiments.
  mean.clonal <- cbind(mean.clonal, exp.mean.clonal)
  mean.nonclonal <- cbind(mean.nonclonal, exp.mean.nonclonal)
  var.clonal <- cbind(var.clonal, exp.var.clonal)
  var.nonclonal <- cbind(var.nonclonal, exp.var.nonclonal)
}

# Check gene name order is consistent across matrices.
all.equal(rownames(mean.clonal), rownames(mean.nonclonal))
all.equal(rownames(var.clonal), rownames(var.nonclonal))
all.equal(rownames(mean.clonal), rownames(var.clonal))

# Plot mean vs variance of gene expression in clonal - nonclonal.
title = paste0(Tcelltype, "+ T cells: ", ncol(mean.clonal), " clonotypes")
diff <- data.frame(gene = rownames(gcdata.sub.matrix))
diff$mean <- rowMeans(mean.clonal) - rowMeans(mean.nonclonal)
diff$var <- rowMeans(var.clonal) - rowMeans(var.nonclonal)
diff.subset <- filter(diff, var^2+mean^2 > 2)  # label cells that are most distant from origin
ggplot(data = diff) + geom_point(mapping = aes(x = var, y = mean)) + geom_text(data = diff.subset, mapping = aes(x = var, y = mean, label = gene), size = 3, nudge_x = 0.3) + labs(title = title, x = "Variability (clonally - nonclonally expanded)", y = "Expression (clonally - nonclonally expanded)") + geom_point(data = filter(diff, gene == "KLRB1"), mapping = aes(x = var, y = mean), col = "red")
ggsave(paste0(figures.dir, "variability_expression_", Tcelltype, ".png"), width = 6, height = 6)
```

## Mean and variance of gene expression in nonclonal CD4 and CD8 T cell populations
Look at gene expression in nonclonal cells.
```{r mean_var}
# Settings for TCR TraCeR analysis.
Tcelltype <- list("CD4" = c("CD4+"), "Treg" = c("Treg"), "CD4_Treg" = c("CD4+", "Treg"), "CD8" = c("CD8+"), "CD8NK" = c("CD8+ with NK markers"), "CD8_CD8NK" = c("CD8+", "CD8+ with NK markers"))

cell.type <- "CD4"
cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$clonal.CD4 == 2, ])
# cell.type <- "CD8"
# cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$clonal.CD8 == 2, ])

# Mean vs variance of gene expression in all nonclonal cells.
mean.cells <- apply(X = gcdata.sub@data[, cells], MARGIN = 1, FUN = function(x) mean(expm1(x)))
sd.cells <- apply(X = gcdata.sub@data[, cells], MARGIN = 1, FUN = function(x) sd(expm1(x)))
mean.cells <- log2(mean.cells + 1)  # return to log space
sd.cells <- log2(sd.cells + 1)  # return to log space

df <- data.frame(mean = mean.cells, sd = sd.cells)
p1 <- ggplot(data = df) + geom_point(mapping = aes(mean, sd), size = 0.5, alpha = 1/10) + labs(x = "mean expression", y = "standard deviation") + geom_point(data = df["KLRB1",], mapping = aes(mean, sd), col = "red")
ggsave(paste0(figures.dir, "mean_sd_expression_allexps_", cell.type, ".png"), width = 6, height = 8)


# # Mean vs variance of gene expression calculated within each experiment, and then averaged.
# # Initialize matrix that will store mean and standard deviation of gene expression within each experiment.
# # Each experiment is a row and each column is the mean or standard deviation of gene expression within
# # that experiment.
# mean.cells.exp <- matrix(nrow = n_distinct(annotation$experiment), ncol = nrow(gcdata.sub@data))
# sd.cells.exp <- matrix(nrow = n_distinct(annotation$experiment), ncol = nrow(gcdata.sub@data))
# rownames(mean.cells.exp) <- levels(annotation$experiment)
# rownames(sd.cells.exp) <- levels(annotation$experiment)
# colnames(mean.cells.exp) <- rownames(gcdata.sub@data)
# colnames(sd.cells.exp) <- rownames(gcdata.sub@data)

# # Iterate over experiments and calculate mean and standard deviation of gene expression within that experiment.
# for (exp in levels(annotation$experiment)) {
#   # Subset on experiment and cell type.
#   cells <- rownames(gcdata.sub@meta.data[gcdata.sub@meta.data$experiment == exp & gcdata.sub@meta.data$annotate %in% Tcelltype[[cell.type]] & rownames(gcdata.sub@meta.data) %in% nonclonal, ]) 
#   
#   mean.cells <- apply(X = gcdata.sub@data[, cells], MARGIN = 1, FUN = function(x) mean(expm1(x)))
#   sd.cells <- apply(X = gcdata.sub@data[, cells], MARGIN = 1, FUN = function(x) sd(expm1(x)))
#   mean.cells.exp[exp,] <- log2(mean.cells + 1)  # return to log space
#   sd.cells.exp[exp,] <- log2(sd.cells + 1)  # return to log space
# }
# # Take mean of expression values across each experiment.
# # mean.cells <- colMeans(mean.cells.exp)
# # sd.cells <- colMeans(sd.cells.exp)
# mean.cells <- colMeans(2^mean.cells.exp-1)
# sd.cells <- colMeans(2^sd.cells.exp-1)
# mean.cells <- log2(mean.cells + 1)  # return to log space
# sd.cells <- log2(sd.cells + 1)  # return to log space
# 
# df <- data.frame(mean = mean.cells, sd = sd.cells)
# p1 <- ggplot(data = df) + geom_point(mapping = aes(mean, sd), size = 0.5, alpha = 1/10) + labs(x = "mean expression", y = "standard deviation") + geom_point(data = df["KLRB1",], mapping = aes(mean, sd), col = "red") + labs(title = "")
# ggsave(paste0(figures.dir, "mean_sd_expression_individualexps_", cell.type, ".png"), width = 6, height = 8)

```
