---
title: "SS2 TCR analysis"
output: html_document
---

Processing of Tracer results for TCR analysis of SS2 data. Adapted from code from Orr Ashenberg. Add results to processed seurat object 


```{r}
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
library(Seurat)
library(dplyr)
library(Matrix)
library(GSA)
library(ggbeeswarm)

base_dir = "/Users/jlabelle/"
script_dir = paste0(base_dir,"Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Shared/Scripts/")
working_dir<- paste0(base_dir, 
                     "Dropbox (Partners HealthCare)/FilbinLab/",
                     "data_analysis/Projects/git_repos/Pedi_Immune/ImmuneCellAnalysis/01_Preprocessing/01_Tracer_Preprocess/")

analysis_dir<- paste0(working_dir, "/analysis/")
figure_dir<- paste0(working_dir, "/figures/")

if(!dir.exists(figure_dir)){dir.create(figure_dir, recursive = TRUE)}
if(!dir.exists(analysis_dir)){dir.create(analysis_dir, recursive = TRUE)}

preprocessing = paste0(script_dir, "single_cell_preprocessing_helper_functions.R")
nmf_helper = paste0(script_dir, "NMF_helper_function.R")
source(preprocessing)
source(nmf_helper)
source(paste0(script_dir, "/plotutils.R"))
source(paste0(script_dir, "seurat3utils.R"))
source(paste0(script_dir, "color.R"))

```

## Load in seurat object
Perform clonal processing using full immune seurat object. Once complete, add to final tcell/cd4/cd8 seurat objects (end of code)
```{r}
seurat_obj<- readRDS("../02_Counts_Preprocess/analysis/broad_annotation/pedOnly_nomuv63/tcell.myeloid_seurat.Rds")

## rename any old clonality columns- want to not overwrite
seurat_obj$clonal_original<- seurat_obj$clonal; seurat_obj$clonal<-NULL
```


## Initialize empty data frames / vectors for storing clonality information
```{r}
## Get alll samples with tracer results
experiment <- list.files("data")
experiment<- experiment[!grepl("ARCHIVE", experiment)]

## Rename to match seurat object- all samples should be included
experiment<- gsub("MUV", "MUV0", experiment)
seurat_use<- subset(seurat_obj, sampleid %in% experiment)
experiment<- experiment[experiment %in% seurat_use$sampleid]
experiment<- factor(experiment, levels=experiment)
sample_cells<- as.character(seurat_use$sampleid)

# Initialize list of clonally expanded T cells keyed by experiment patient sample that the cells come from.
# clonal is list containing the name of the clonotype cells in each experiment.
# clonotypes.cells is list containing all the metadata for each clonotype cell in each experiment.
# each experiment is a list of dataframes, and each dataframe is a clonal group of cells from that experiment
clonal <- vector("list", length = length(sample_cells))
clonotypes.cells <- vector("list", length = length(sample_cells))  
names(clonal) <- sample_cells
names(clonotypes.cells) <- sample_cells

# Make a dataframe containing cells from all clonotypes across all experiments that pass quality control. 
# Clonal groups get renumbered to ensure unique clonotype group numbers across all experiments.
clonotypes.all <- data.frame(cell_name = character(), experiment = character(), 
                             clonal_group = vector(), group_size = vector(), stringsAsFactors = F)

# offset gets updated after each experiment to ensure unique group number for clonotypes
group.offset <- 0  

# Store a vector of number of T cells found in each clonotype.
clonotypes.size <- vector()  

# Store a list where each element is keyed by experiment and contains a vector of the names of cells 
# that have a reconstructed TCR.
tcell.reconstructed <- vector("list", length = length(sample_cells))

# Experiment id, number of cells passing QC, number of cells with productive A or B chain, 
# number of cells with productive A and B chains, 
# number of clonally expanded T cells, number of T cell clonotypes
df.plot <- data.frame(experiment = character(), nExps = vector(), nCells = vector(), productiveAB = vector(), 
                      nExpanded = vector(), nClonotypes = vector())
```


```{r}
for (exp in as.character(experiment)) {
  exp.filename <- exp
  print(exp)
  file.celldata <- paste0(working_dir, "/data/",gsub("MUV0", "MUV", exp.filename), "/output/filtered_TCRAB_summary/cell_data.csv")
  print(file.exists(file.celldata))
}
```

## TraCeR T-cell receptor reconstruction statistics
We look at number of total cells (some cells may be missing TCR reads), number of productive A chains, productive B chains, number of productive A and B chains, number of expanded T cells in clonotypes, and number of clonotypes.
```{r tracer_stats}
## Cycle through each sample, reading in tracer results and exporting stats
for (exp in as.character(experiment)) {
  exp.filename <- exp
  print(exp)
  file.celldata <- paste0(working_dir, "/data/",gsub("MUV0", "MUV", exp.filename), "/output/filtered_TCRAB_summary/cell_data.csv")
  
  # If cell_data file doesn't exist, that means no TCR chains were reconstructed. 
  # Collect basic stats and skip to next sample
  if (!file.exists(file.celldata)) {
    tcell.reconstructed[[exp]] <- character()
    n.exps <- filter(seurat_use[[]], sampleid == exp) %>% nrow()  # T cells in this experiment passing quality control
    n.cells <- 0  # cells in this experiment with productive A chain or productive B chain
    n.productiveAB <- 0  # productive A chain and productive B chain
    n.expanded <- 0  # expanded T cells
    n.clonotypes <- 0  # clonotypes
    df.plot <- rbind(df.plot, data.frame(exp, n.exps, n.cells, n.productiveAB, n.expanded, n.clonotypes))
    next 
  }
  
  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(file.celldata, header = T, sep = ",", stringsAsFactors = F)
  tcr.annotate$cell_name<- gsub("MUV", "MUV0", tcr.annotate$cell_name)
  
  # Rename cell names as needed to match seurat_obj
  tcr.annotate$cell_name<- gsub("\\.", "_", tcr.annotate$cell_name)

  # Identify all T cells in this experiment containing a reconstructed TCR chain A or chain B.
  tcell.reconstructed[[exp]] <- tcr.annotate[!(tcr.annotate$A_productive=="") | !(tcr.annotate$B_productive==""), "cell_name"]
    
  # Identify clonally expanded T cells.
  if (!all(is.na(tcr.annotate$clonal_group))) {  
    label.clonotypes <- c(0:max(tcr.annotate$clonal_group, na.rm = T))  # clonotype number labels start at 0
    clonotypes <- filter(tcr.annotate, clonal_group %in% label.clonotypes)  # T cells that are clonally expanded
  } else {
    clonotypes <- data.frame()
  }
  
  # Collect stats on clonally expanded T cells.
    # T cells in this experiment passing quality control
    n.exps <- filter(seurat_use[[]], sampleid == exp) %>% nrow()  
    
    # cells in this experiment with productive A chain or productive B chain
   n.cells <- tcr.annotate[!(tcr.annotate$A_productive=="") | !(tcr.annotate$B_productive==""),] %>% nrow  
   
   # productive A chain and productive B chain
    n.productiveAB <- tcr.annotate[!(tcr.annotate$A_productive=="") & !(tcr.annotate$B_productive==""),] %>% nrow  
    
    # expanded T cells
    n.expanded <- nrow(clonotypes)  
    
    # clonotypes
    n.clonotypes <- length(unique(clonotypes$clonal_group))  
    
    df.plot <- rbind(df.plot, data.frame(exp, n.exps, n.cells, n.productiveAB, n.expanded, n.clonotypes))
  
  # Keep clonally expanded T cells that are also found in Tcell gene expression matrix and not filtered out.
  # Names in gene expression matrix should match those in TraCeR data
  clonal[[exp]] <- intersect(clonotypes$cell_name, colnames(seurat_use))
  remove.cells <- setdiff(clonotypes$cell_name, clonal[[exp]])
  clonotypes <- clonotypes[!clonotypes$cell_name %in% remove.cells, ]
  if (nrow(clonotypes)) {
    clonotypes <- clonotypes %>% group_by(clonal_group) %>% mutate(group_size = length(clonal_group))  # update number of cells in each clonotype after having removed cells that were not found in expression matrix
    clonotypes <- clonotypes[clonotypes$group_size > 1, ]  # remove clonotypes that now only have 1 cell
  }
  print(c(length(clonal[[exp]]), n.expanded))
  if (length(remove.cells)) {print(remove.cells)}  # cells that were likely filtered from expression matrix

  # For each clonotype, store the cells that make up that clonotype.  
  clonotypes.cells[[exp]] <- split(clonotypes, clonotypes$clonal_group)

  # Store size of each clonotype.
  size <- sapply(clonotypes.cells[[exp]], function(clone) unique(clone$group_size)) %>% unname()
  if (length(size)) {
    clonotypes.size <- c(clonotypes.size, size)
    print(paste0("clonotype sizes: ", paste(size, collapse = " ")))
  }
  
  # Place cells for all clonotypes across all experiments in a single dataframe, with each clonotype 
  # group receiving a unique number. Within each experiment, clonal_group numbering starts at 0.
  if (nrow(clonotypes)) {
    clonotypes <- clonotypes[order(clonotypes$clonal_group),]  # sort by clonal_group
    clonotypes$clonal_group <- clonotypes$clonal_group + group.offset
    clonotypes[["experiment"]] <- exp 
    clonotypes$cell_name <- as.character(clonotypes$cell_name)  # remove factor levels from cell names
    clonotypes.all <- bind_rows(clonotypes.all, clonotypes[, c("cell_name", "experiment", "clonal_group", "group_size")])
    group.offset <- max(clonotypes$clonal_group) + 1  # update clonal_group number for next experiment in iteration of this loop
  }
}

```

## Plot clonotype/cell reconstruction results
```{r}
# Set order of experiments when plotting
df.plot$exp <- factor(df.plot$exp, levels = experiment)

# Gather df.plot dataframe into stat : count mapping key-value pairs for making bar plots. 
labels.plot <- c("number of cells", "productive A or B chain", "productive A and B chain", "T cells with recurrent TCRs", "clonotypes")
bar.plot <- gather(df.plot, "stat", 'count', c(2:6))

  # Set order in which different statistics from TraCer are plotted.
  bar.plot$stat <- factor(bar.plot$stat, levels = colnames(df.plot)[2:6])
  ggplot(data = bar.plot) + 
    geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = position_dodge()) + 
    labs(x = "experiment", y = "count") + 
    theme(axis.text.x  = element_text(angle=90)) 
  
  ggsave(paste0(figure_dir, 'tracer_stats.png'), width = 6, height = 6)

# Graph number of cells passing QC and number of productive T cells per patient
labels.plot <- c("Total T cells", bquote("TCR" ~ alpha ~ "or" ~ beta))
bar.plot <- gather(df.plot, "stat", 'count', c(2:3))

  # Set order in which different statistics from TraCer are plotted.
  bar.plot$stat <- factor(bar.plot$stat, levels = colnames(df.plot)[2:3])
  
  ggplot(data = bar.plot) + 
    geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = "identity") + 
    labs(x = "Sample", y = "Number of cells") + theme(axis.text.x  = element_text(angle=90)) 
  ggsave(paste0(figure_dir, 'tracer_ncells_nproductivechain.png'), width = 8, height = 6)

# Graph number of clonotypes per patient.
ggplot(data = df.plot[, c("exp", "n.clonotypes")]) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = n.clonotypes)) + 
  labs(x = "experiment", y = "Number of clonotypes") + 
  theme(axis.text.x  = element_text(angle=90)) 
ggsave(paste0(figure_dir, 'tracer_nclonotypes.png'), width = 6, height = 6)

# Graph percent of T cells with productive A or B chain that are clonally expanded in each patient.
df.plot$percent.expanded <- 100 * df.plot$n.expanded / df.plot$n.cells
ggplot(data = df.plot[, c("exp", "percent.expanded")]) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = percent.expanded)) + 
  labs(x = "experiment", y = "T cells with recurrent TCRs (%)") + 
  theme(axis.text.x  = element_text(angle=90)) + ylim(c(0, 50))
ggsave(paste0(figure_dir, 'tracer_expanded.png'), width = 6, height = 6)

# Scatter plot of T cells with recurrent TCRS (%) vs Number of clonotypes.
ggplot(data = df.plot, mapping = aes(x = n.clonotypes, y = percent.expanded, label = exp)) + 
  geom_point(mapping = aes(size = n.cells)) + 
  geom_text(nudge_y = 2) + 
  labs(x = "Number of clonotypes", y = "T cells with recurrent TCRs (%)") + 
  scale_size_continuous(name = bquote(atop("Number of TCR", ~ alpha ~ "or" ~ beta))) + 
  ylim(0, 52)
ggsave(paste0(figure_dir, 'tracer_expanded_vs_nproductivechain.png'), width = 8, height = 6)
cor(x = df.plot$n.clonotypes, y = df.plot$percent.expanded, method = "pearson")

# Distribution of clonotype sizes in each experiment. 
# These plots show both the clonotype sizes, and the sizes of the clonotypes normalized by the 
# number of cells in that experiment for which there was a productive A chain or B chain.
df.size <- clonotypes.all %>% group_by(experiment, clonal_group) %>% summarise(size = unique(group_size))
df.size$experiment <- factor(df.size$experiment, levels = levels(experiment))
df.size$size.normalized <- 0  # normalize clonotype size by number of cells
for (exp in unique(experiment)) {
  df.size[df.size$experiment == exp, "size.normalized"] <- 100 * df.size[df.size$experiment == exp, "size"] / 
    df.plot[df.plot$exp == exp, "n.cells"]
}
ggplot(data = df.size, mapping = aes(x = experiment, y = size)) + 
  geom_quasirandom(groupOnX = T, cex = 1, varwidth = T) + 
  labs(y = "clonotype size") + scale_y_continuous(trans = "log2")
ggsave(paste0(figure_dir, "fig2_tracer_clonotype_size_beeswarm.png"), width = 15, height = 6)


ggplot(data = df.size, mapping = aes(x = experiment, y = size.normalized)) + 
  geom_quasirandom(groupOnX = T, cex = 1, varwidth = T) + 
  labs(y = "percent of reconstructed T cells in clonotype")
ggsave(paste0(figure_dir, "tracer_clonotype_sizenormalized_beeswarm.png"), width = 15, height = 6)

# Histogram of clonotype sizes (number of T cells in a clonotype).
ggplot(data = data.frame(size = clonotypes.size)) + 
  geom_histogram(mapping = aes(x = size), binwidth = 1, color="black", fill="white") + 
  scale_x_continuous(breaks=seq(0,60,5)) + 
  scale_y_continuous(expand = c(0,0)) + 
  labs(x = "clonotype size", y = "count") 
ggsave(paste0(figure_dir, 'tracer_clonotype_histogram.png'), width = 8, height = 6)

# Write a file summarizing all clonotypes.
rownames(clonotypes.all) <- clonotypes.all$cell_name
write.csv(clonotypes.all, file=paste0(analysis_dir, "clonotypes_all.csv"), quote = F)

## Create df with all clonal info for batch 2- same as for batch 1


# Save current progress.
save( clonal, clonotypes.cells, clonotypes.all, tcell.reconstructed, file = paste0(analysis_dir, "clonalTcells_cluster.Rda"))

```

## Load in results
```{r}
load(paste0(analysis_dir, "clonalTcells_cluster.Rda"))

## initialize clonality dataframe for these samples
clonal<- seurat_obj@meta.data[seurat_obj$sampleid %in% names(tcell.reconstructed),c("orig.ident", "sampleid") ]

## Add whether TCR was reconstructed
tcell.reconstructed<- unlist(unname(tcell.reconstructed))
clonal$tcr_reconstructed<- ifelse(rownames(clonal) %in% tcell.reconstructed,"Yes", "No")

## Add clonal group and group size
clonal<- merge(clonal, clonotypes.all[,c("clonal_group", "group_size")], by=0, all.x=TRUE)
rownames(clonal)<-clonal$Row.names; clonal$Row.names<- NULL

## Fix group size- if tcr is reconstructed but not clonally expanded (e.g., clonal group ==NA), set group size to 1
clonal$group_size<- ifelse(clonal$tcr_reconstructed=="Yes" & is.na(clonal$clonal_group), 1, clonal$group_size)

## Add clonality
clonal$clonal<- ifelse(!is.na(clonal$clonal_group), "clonal", "nonclonal")
clonal<- clonal[order(clonal$sampleid, clonal$clonal_group),]

write.csv(clonal, file=paste0(analysis_dir, "clonal_stats.csv"))
```


## Add clonality results to seurat object 

```{r}
## Load clonal results data frame
clonal<- read.csv(paste0(analysis_dir, "clonal_stats.csv"), row.names = "X")

## Load seurat objects- immune, tcell, cd4, cd8
tcell_data_dir<- "../../02a_Tcells/02_Ped.Adult/03_Visualize.Reannotate_programs/analysis/Reannotate_programs/"
tcell_data_dir<- paste0(tcell_data_dir, "rank8_sharedByunion/")
immune_seurat<- readRDS("../02_Counts_Preprocess/analysis/broad_annotation/pedOnly_nomuv63/tcell.myeloid_seurat.Rds")
tcell_seurat<- readRDS(paste0(tcell_data_dir, "ped_seurat_tcell_harmony.Rds"))
cd4_seurat<- readRDS(paste0(tcell_data_dir, "ped_seurat_cd4_harmony.Rds"))
cd8_seurat<- readRDS(paste0(tcell_data_dir, "ped_seurat_cd8_harmony.Rds"))

## move any current clonal info to "original" column
my_renameClonalColumns<- function(seurat){
  seurat$clonal_original<- seurat$clonal
  seurat$clonal_group_original<- seurat$clonal_group
  seurat$tcr_reconstructed_original<- seurat$tcr_reconstructed
  seurat$group_size_original<- seurat$group_size
  return(seurat)
}
tcell_seurat<- my_renameClonalColumns(tcell_seurat)
cd4_seurat<- my_renameClonalColumns(cd4_seurat)
cd8_seurat<- my_renameClonalColumns(cd8_seurat)


## Add onto seurat objects
immune_seurat<- AddMetaData(immune_seurat, clonal[,c("tcr_reconstructed", "clonal_group", "group_size", "clonal")] )
tcell_seurat<- AddMetaData(tcell_seurat, clonal[,c("tcr_reconstructed", "clonal_group", "group_size", "clonal")] )
cd4_seurat<- AddMetaData(cd4_seurat, clonal[,c("tcr_reconstructed", "clonal_group", "group_size", "clonal")] )
cd8_seurat<- AddMetaData(cd8_seurat, clonal[,c("tcr_reconstructed", "clonal_group", "group_size", "clonal")] )

## save
saveRDS(immune_seurat, file="../02_Counts_Preprocess/analysis/broad_annotation/pedOnly_nomuv63/tcell.myeloid_seurat.Rds")
saveRDS(tcell_seurat, file=paste0(tcell_data_dir, "ped_seurat_tcell_harmony.Rds"))
saveRDS(cd4_seurat, file=paste0(tcell_data_dir, "ped_seurat_cd4_harmony.Rds"))
saveRDS(cd8_seurat, file=paste0(tcell_data_dir, "ped_seurat_cd8_harmony.Rds"))
```