---
title: "PostTracer_Processing"
author: "Orr Ashenberg"
date: "11/04/21"
output: html_document
---
This script analyzes the T cells from glioblastoma samples. It takes the T cells clustered by the script ss2_pedbrain_cluster.Rmd, and loads that Seurat object. The T cells have already had quality control filtering for genes and cells, and clustering performed. Here, we can do follow up TCR and TraCeR analysis.

Run the first two chunks of this markdown file, before running any other chunks.

```{r init}
# Load libraries
rm(list = ls())
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())
library(Seurat)
library(dplyr)
library(Matrix)
library(GSA)
library(ggbeeswarm)

# Set directories
user.path <- "/Users/jlabelle/"
proj.path <-paste0(user.path, 
                     "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Projects/git_repos/Pedi_Immune/ImmuneCellAnalysis/01_Preprocessing/",
                   "01_Tracer_Preprocess/")

name.project <- "pedbrain_immune_newSamples"


# Load T cell Seurat object generated by ss2_pedbrain_cluster.Rmd.
# load("/Users/orr/Documents/Rdata/2020_12_10_pedbrain_immune_no_batchcorrect.Rda") 
Rda.nobatchcorrect.path <- paste0(proj.path, "../02_Counts_Preprocess/analysis/newSamples/gcdata_filtered.Rds")
gcdata<- readRDS(Rda.nobatchcorrect.path)
gcdata$sampleid<- gsub("MUV0", "MUV", gcdata$sampleid)

# Set project variables using annotation.
experiment <- factor(unique(gcdata$sampleid), levels = unique(gcdata$sampleid)) 
sampleid <- unique(gcdata$sampleid)

# Settings for T cell types.
Tcelltype <- "CD8"
types <- list("CD4" = c("CD4 T"), "Treg" = c("Treg"), "CD4_Treg" = c("CD4 T",  "Treg"), "CD8" = c("CD8 T"))
cell.type <- types[[Tcelltype]]

# Order in which to display annotated T cell subsets when plotting.
group.order <- c("CD8 T", "CD4 T", "Treg", "Cycling T cells", "Mono")

# Set minimum size of clonotype to consider.
group_size.cutoff <- 2  # need at least this many cells to define a clonotype
# Number of iterations of choosing random sets of cells in nonclonal gene expression analysis.
# n.runs <- 5
n.runs <- 10

# Output figures and Broad single-cell portal
figures.dir <- paste0(proj.path, "/figures/")    
figures.path <- sapply(sampleid, function(i) paste0(figures.dir, i, "_"), USE.NAMES = F) 
portal.path <- paste0(figures.dir, "/portal")
Rda.clonalTcells.path<- paste0("analysis/clonalTcells_cluster.Rda")

# Save seurat object expression matrix data, tSNE, clustering, and cluster annotation scores.
Rda.clonalTcells.path <- paste0(proj.path, "/analysis/clonalTcells_cluster.Rda")
#Rda.clonalTcells.NMF.path <- paste0(proj.path, "/src/Rdata/", date, "_clonalTcells_NMF_cluster.Rda")

# Read in a list of cell cycle markers, from Tirosh et al, 2015.
# We can segregate this list into markers of G2/M phase and markers of S phase.
cc.genes <- read.table(paste0(user.path,"Dropbox (Partners HealthCare)/FilbinLab/data_analysis/",
                              "Marker_genes/regev_lab_cell_cycle_genes.txt"))$V1
s.genes <- cc.genes[1:43]
g2m.genes <- cc.genes[44:97]


# Load my functions.
source(paste0(user.path, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/", "/Scripts/plotutils.R"))
source(paste0(user.path, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/", "/Scripts/seurat3utils.R"))
source(paste0(user.path, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/", "/Scripts/color.R"))
```



## TraCeR T-cell receptor reconstruction statistics
We look at number of total cells (some cells may be missing TCR reads), number of productive A chains, productive B chains, number of productive A and B chains, number of expanded T cells in clonotypes, and number of clonotypes.
```{r tracer_stats}
run.exps <- as.character(sampleid)

# List of clonally expanded T cells keyed by experiment patient sample that the cells come from.
# clonal is list containing the name of the clonotype cells in each experiment.
# clonotypes.cells is list containing all the metadata for each clonotype cell in each experiment.
clonal <- vector("list", length = length(run.exps))
clonotypes.cells <- vector("list", length = length(run.exps))  # each experiment is a list of dataframes, and each dataframe is a clonal group of cells from that experiment
names(clonal) <- run.exps
names(clonotypes.cells) <- run.exps 

# Make a dataframe containing cells from all clonotypes across all experiments that pass quality control. 
# Clonal groups get renumbered to ensure unique clonotype group numbers across all experiments.
clonotypes.all <- data.frame(cell_name = character(), experiment = character(), clonal_group = vector(), group_size = vector(), stringsAsFactors = F)
group.offset <- 0  # offset gets updated after each experiment to ensure unique group number for clonotypes

# Store a vector of number of T cells found in each clonotype.
clonotypes.size <- vector()  

# Store a list where each element is keyed by experiment and contains a vector of the names of cells 
# that have a reconstructed TCR.
tcell.reconstructed <- vector("list", length = length(run.exps))

# Experiment id, number of cells passing QC, number of cells with productive A or B chain, 
# number of cells with productive A and B chains, 
# number of clonally expanded T cells, number of T cell clonotypes
df.plot <- data.frame(experiment = character(), nExps = vector(), nCells = vector(), productiveAB = vector(), 
                      nExpanded = vector(), nClonotypes = vector())
run.exps<- c("BT1857", "BT1873", "MUV92", "MUV94", "MUV95")

for (exp in run.exps) {
  exp.filename <- exp
  print(exp)
  file.celldata <- paste0(proj.path, "/data/", exp.filename, "/output/filtered_TCRAB_summary/cell_data.csv")
  
  # If cell_data file doesn't exist, that means no TCR chains were reconstructed. 
  # Collect basic stats and skip to next sample
  if (!file.exists(file.celldata)) {
    tcell.reconstructed[[exp]] <- character()
    n.exps <- filter(gcdata[[]], sampleid == exp) %>% nrow()  # T cells in this experiment passing quality control
    n.cells <- 0  # cells in this experiment with productive A chain or productive B chain
    n.productiveAB <- 0  # productive A chain and productive B chain
    n.expanded <- 0  # expanded T cells
    n.clonotypes <- 0  # clonotypes
    df.plot <- rbind(df.plot, data.frame(exp, n.exps, n.cells, n.productiveAB, n.expanded, n.clonotypes))
    next 
  }
  
  # Read in TCR annotation file containing TraCeR information on each experimental sample.
  tcr.annotate <- read.table(file.celldata, header = T, sep = ",", stringsAsFactors = F)
  
  # Rename cell names as needed to match gcdata
  tcr.annotate$cell_name<- gsub("_", "\\.", gsub("highdepth.", "", tcr.annotate$cell_name))

  # Identify all T cells in this experiment containing a reconstructed TCR chain A or chain B.
  tcell.reconstructed[[exp]] <- tcr.annotate[!(tcr.annotate$A_productive=="") | !(tcr.annotate$B_productive==""), "cell_name"]
    
  # Identify clonally expanded T cells.
  if (!all(is.na(tcr.annotate$clonal_group))) {  # fix this line
    label.clonotypes <- c(0:max(tcr.annotate$clonal_group, na.rm = T))  # clonotype number labels start at 0
    clonotypes <- filter(tcr.annotate, clonal_group %in% label.clonotypes)  # T cells that are clonally expanded
  } else {
    clonotypes <- data.frame()
  }
  
  # Collect stats on clonally expanded T cells.
  n.exps <- filter(gcdata[[]], sampleid == exp) %>% nrow()  # T cells in this experiment passing quality control
  n.cells <- tcr.annotate[!(tcr.annotate$A_productive=="") | !(tcr.annotate$B_productive==""),] %>% nrow  # cells in this experiment with productive A chain or productive B chain
  n.productiveAB <- tcr.annotate[!(tcr.annotate$A_productive=="") & !(tcr.annotate$B_productive==""),] %>% nrow  # productive A chain and productive B chain
  n.expanded <- nrow(clonotypes)  # expanded T cells
  n.clonotypes <- length(unique(clonotypes$clonal_group))  # clonotypes
  df.plot <- rbind(df.plot, data.frame(exp, n.exps, n.cells, n.productiveAB, n.expanded, n.clonotypes))
  
  # Keep clonally expanded T cells that are also found in Tcell gene expression matrix and not filtered out.
  # Names in gene expression matrix should match those in TraCeR data (E30TC_P5_E12_E30_CD4+CD8+_P5P6).
  clonal[[exp]] <- intersect(clonotypes$cell_name, colnames(gcdata))
  remove.cells <- setdiff(clonotypes$cell_name, clonal[[exp]])
  clonotypes <- clonotypes[!clonotypes$cell_name %in% remove.cells, ]
  if (nrow(clonotypes)) {
    clonotypes <- clonotypes %>% group_by(clonal_group) %>% mutate(group_size = length(clonal_group))  # update number of cells in each clonotype after having removed cells that were not found in expression matrix
    clonotypes <- clonotypes[clonotypes$group_size > 1, ]  # remove clonotypes that now only have 1 cell
  }
  print(c(length(clonal[[exp]]), n.expanded))
  if (length(remove.cells)) {print(remove.cells)}  # cells that were likely filtered from expression matrix

  # For each clonotype, store the cells that make up that clonotype.  
  clonotypes.cells[[exp]] <- split(clonotypes, clonotypes$clonal_group)

  # Store size of each clonotype.
  size <- sapply(clonotypes.cells[[exp]], function(clone) unique(clone$group_size)) %>% unname()
  if (length(size)) {
    clonotypes.size <- c(clonotypes.size, size)
    print(paste0("clonotype sizes: ", paste(size, collapse = " ")))
  }
  
  # Place cells for all clonotypes across all experiments in a single dataframe, with each clonotype 
  # group receiving a unique number. Within each experiment, clonal_group numbering starts at 0.
  if (nrow(clonotypes)) {
    clonotypes <- clonotypes[order(clonotypes$clonal_group),]  # sort by clonal_group
    clonotypes$clonal_group <- clonotypes$clonal_group + group.offset
    clonotypes[["experiment"]] <- exp 
    clonotypes$cell_name <- as.character(clonotypes$cell_name)  # remove factor levels from cell names
    clonotypes.all <- bind_rows(clonotypes.all, clonotypes[, c("cell_name", "experiment", "clonal_group", "group_size")])
    group.offset <- max(clonotypes$clonal_group) + 1  # update clonal_group number for next experiment in iteration of this loop
  }
}

```

## Plot clonotype/cell reconstruction results
```{r}
# Set order of experiments when plotting
df.plot$exp <- factor(df.plot$exp, levels = sampleid)

# Gather df.plot dataframe into stat : count mapping key-value pairs for making bar plots. 
labels.plot <- c("number of cells", "productive A or B chain", "productive A and B chain", "T cells with recurrent TCRs", "clonotypes")
bar.plot <- gather(df.plot, "stat", 'count', c(2:6))
# Set order in which different statistics from TraCer are plotted.
bar.plot$stat <- factor(bar.plot$stat, levels = colnames(df.plot)[2:6])
ggplot(data = bar.plot) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = position_dodge()) + 
  labs(x = "experiment", y = "count") + 
  theme(axis.text.x  = element_text(angle=90)) 

ggsave(paste0(figures.dir, 'tracer_stats.png'), width = 6, height = 6)

# Graph number of cells passing QC and number of productive T cells per patient
labels.plot <- c("Total T cells", bquote("TCR" ~ alpha ~ "or" ~ beta))
bar.plot <- gather(df.plot, "stat", 'count', c(2:3))
# Set order in which different statistics from TraCer are plotted.
bar.plot$stat <- factor(bar.plot$stat, levels = colnames(df.plot)[2:3])
# ggplot(data = bar.plot) + geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = position_dodge()) + scale_fill_manual(values = material.heat(2), name="", breaks=levels(bar.plot$stat), labels=labels.plot) + labs(x = "experiment", y = "Number of cells") + theme(axis.text.x  = element_text(angle=90)) 
ggplot(data = bar.plot) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = count, fill = stat), position = "identity") + 
  labs(x = "Sample", y = "Number of cells") + theme(axis.text.x  = element_text(angle=90)) 
ggsave(paste0(figures.dir, 'fig2_tracer_ncells_nproductivechain.png'), width = 8, height = 6)

# Graph number of clonotypes per patient.
ggplot(data = df.plot[, c("exp", "n.clonotypes")]) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = n.clonotypes)) + 
  labs(x = "experiment", y = "Number of clonotypes") + 
  theme(axis.text.x  = element_text(angle=90)) 
ggsave(paste0(figures.dir, 'fig2_tracer_nclonotypes.png'), width = 6, height = 6)

# Graph percent of T cells with productive A or B chain that are clonally expanded in each patient.
df.plot$percent.expanded <- 100 * df.plot$n.expanded / df.plot$n.cells
ggplot(data = df.plot[, c("exp", "percent.expanded")]) + 
  geom_bar(stat = "identity", mapping = aes(x = exp, y = percent.expanded)) + 
  labs(x = "experiment", y = "T cells with recurrent TCRs (%)") + 
  theme(axis.text.x  = element_text(angle=90)) + ylim(c(0, 50))
ggsave(paste0(figures.dir, 'fig2_tracer_expanded.png'), width = 6, height = 6)

# Scatter plot of T cells with recurrent TCRS (%) vs Number of clonotypes.
ggplot(data = df.plot, mapping = aes(x = n.clonotypes, y = percent.expanded, label = exp)) + 
  geom_point(mapping = aes(size = n.cells)) + 
  geom_text(nudge_y = 2) + 
  labs(x = "Number of clonotypes", y = "T cells with recurrent TCRs (%)") + 
  scale_size_continuous(name = bquote(atop("Number of TCR", ~ alpha ~ "or" ~ beta))) + 
  ylim(0, 52)
ggsave(paste0(figures.dir, 'fig2_tracer_expanded_vs_nproductivechain.png'), width = 8, height = 6)
cor(x = df.plot$n.clonotypes, y = df.plot$percent.expanded, method = "pearson")

# Distribution of clonotype sizes in each experiment. 
# These plots show both the clonotype sizes, and the sizes of the clonotypes normalized by the 
# number of cells in that experiment for which there was a productive A chain or B chain.
df.size <- clonotypes.all %>% group_by(experiment, clonal_group) %>% summarise(size = unique(group_size))
df.size$experiment <- factor(df.size$experiment, levels = levels(experiment))
df.size$size.normalized <- 0  # normalize clonotype size by number of cells
for (exp in unique(experiment)) {
  df.size[df.size$experiment == exp, "size.normalized"] <- 100 * df.size[df.size$experiment == exp, "size"] / 
    df.plot[df.plot$exp == exp, "n.cells"]
}
ggplot(data = df.size, mapping = aes(x = experiment, y = size)) + 
  geom_quasirandom(groupOnX = T, cex = 1, varwidth = T) + 
  labs(y = "clonotype size") + scale_y_continuous(trans = "log2")
ggsave(paste0(figures.dir, "fig2_tracer_clonotype_size_beeswarm.png"), width = 15, height = 6)


ggplot(data = df.size, mapping = aes(x = experiment, y = size.normalized)) + 
  geom_quasirandom(groupOnX = T, cex = 1, varwidth = T) + 
  labs(y = "percent of reconstructed T cells in clonotype")
ggsave(paste0(figures.dir, "tracer_clonotype_sizenormalized_beeswarm.png"), width = 15, height = 6)

# Histogram of clonotype sizes (number of T cells in a clonotype).
ggplot(data = data.frame(size = clonotypes.size)) + 
  geom_histogram(mapping = aes(x = size), binwidth = 1, color="black", fill="white") + 
  scale_x_continuous(breaks=seq(0,60,5)) + 
  scale_y_continuous(expand = c(0,0)) + 
  labs(x = "clonotype size", y = "count") 
ggsave(paste0(figures.dir, 'tracer_clonotype_histogram.png'), width = 8, height = 6)

# Write a file summarizing all clonotypes.
rownames(clonotypes.all) <- clonotypes.all$cell_name
write.csv(clonotypes.all, file=paste0(figures.dir, "clonotypes_all.csv"), quote = F)

# Save current progress.
save( clonal, clonotypes.cells, clonotypes.all, tcell.reconstructed, file = Rda.clonalTcells.path)

```


