---
title: "ProjectBroadAnnots.Rmd"
author: "Jenna LaBelle"
date: "10/07/22"
output: html_document
---

## Expression patterns of various genes of interest acros PD1 treated samples/untreated GBM

```{r}
library(Seurat) 
library(ggplot2)
library(ggpubr)

base_dir = "/Users/jlabelle/"
script_dir = "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Scripts/"
working_dir<- paste0(base_dir, 
                     "Dropbox (Partners HealthCare)/FilbinLab/",
                     "data_analysis/Projects/git_repos/Pedi_Immune/PD1_Treat/03_comparisonToAdult/")

analysis_dir<- paste0(working_dir, "/analysis/goi_expression/")
figure_dir<- paste0(working_dir, "/figures/goi_expression/")
data_dir<- paste0(working_dir, "analysis/Merge/")
full_cohort_dir<- "../../ImmuneCellAnalysis/"

if(!dir.exists(figure_dir)){dir.create(figure_dir, recursive = TRUE)}
if(!dir.exists(analysis_dir)){dir.create(analysis_dir, recursive = TRUE)}

preprocessing = paste0(base_dir, script_dir, "single_cell_preprocessing_helper_functions.R")
nmf_helper = paste0(base_dir, script_dir, "NMF_helper_function.R")
source(preprocessing)
source(nmf_helper)


```


## Load seurat objects: pd-1 (processed in 01b_Preprocess_QC_ToSeurat.Rmd) and full immune cohort (both myeloid and t cells)

```{r}
files<- list.files(data_dir)
seurat_list<- lapply(files, function(f){
   readRDS(paste0(data_dir, f))
})
names(seurat_list)<- gsub("_seurat.Rds", "", files)

## Colors 
tcell_colors<- readRDS(paste0(full_cohort_dir, "/plot_colors/Tcell.programs_rank8sharedByunion_reAnnot.Rds"))
myeloid_colors<- readRDS(paste0(full_cohort_dir, "/plot_colors/Myeloid_program_colors.Rds"))
myeloid_colors<- myeloid_colors$rank6
immune_colors<- c(tcell_colors, myeloid_colors)
colors_treatment<- c(PD1="orange", Untreated="purple")

  
## Nature 2020 genesets
nature_2020_genes<- read.csv(paste0(base_dir, "Dropbox (Partners HealthCare)/FilbinLab/data_analysis/Shared/Marker_genes/Immune/Published/",
                                    "Nature2020_genesets.csv"))
colnames(nature_2020_genes)<- c("cytotoxic", "naive", "predysfunctional", "dysfunctional")
nature_2020_genes<- as.list(nature_2020_genes)
nature_2020_genes<- lapply(nature_2020_genes, function(x){x[x!=""]})

```


## Pseudobulked heatmap of expression of GOI (T cells)
```{r}
GOI<- c( "PDCD1", "TIGIT", "CTLA4", "ITGAE", "GZMK", "GNLY","PRF1", "NKG7", "GZMB", "GZMA")

## By sample
all_plots<- lapply(c("cd4", "cd8"), function(x){
  tmp<- seurat_list[[x]]
  tmp$Treatment<- gsub("Treated", "PD1", tmp$Treatment)
  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$Treatment, "_", tmp$Age_Group)
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  pb_center<- NormCenter(pb)$center_data
  
  ## Plot
  myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE,
            orderFactors = c("Ped", "Adult"), orderSubtypes = c("Untreated", "PD1")) +
    scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=2)
ggsave(paste0(figure_dir, "Heatmap_GOI_Tcells.pdf"),device="pdf", width=13, height=6)


## Log FC comparing change in expression in treated/untreated
all_plots<- lapply(c("cd4", "cd8"), function(x){
  tmp<- seurat_list[[x]]
  tmp$Treatment<- gsub("Treated", "PD1", tmp$Treatment)
  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$Treatment, "_", tmp$Age_Group)
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  ## Calculate FC
   fc<- data.frame(Adult=log( (pb$PD1_Adult-pb$Untreated_Adult)/(pb$Untreated_Adult+0.01) + 1 ), 
                  Ped=log( (pb$PD1_Ped-pb$Untreated_Ped)/(pb$Untreated_Ped+0.01) + 1 ),
                  row.names = rownames(pb))

  
  ## Plot
  myHeatmap(fc, GOI=GOI, min.value = -3, max.value = 3, facetWrap = FALSE,
            orderFactors = c("Ped", "Adult")) +
     scale_fill_gradientn(colors=c( "khaki3", "lemonchiffon1","white","palevioletred1", "palevioletred2", "violetred4"))+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=2)
ggsave(paste0(figure_dir, "Heatmap_foldChange_GOI_Tcells.pdf"),device="pdf", width=8, height=4)

## Log FC comparing change in expression in treated/untreated- by program
seurat_use<- seurat_list$cd8

seurat_use$Treatment<- gsub("Treated", "PD1", seurat_use$Treatment)

## Pseudobulk by program/age/treatment, then calculate log FC comparing treated/untreated
programs_use<-unique(seurat_use$Program_projected)
programs_use<- programs_use[programs_use!="Effector.Memory"]
all_fc<- lapply(programs_use, function(program){ ## don't use em- very few cells (only 5 ped UT)
  ## subset to program
  tmp<- subset(seurat_use, Program_projected==program)
  
  ## Pseuodobulk  
  tmp$pb<- paste0(tmp$Treatment, "_", tmp$Age_Group)
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  fc<- data.frame(Adult=log( (pb$PD1_Adult-pb$Untreated_Adult)/(pb$Untreated_Adult+0.01) + 1 ), 
                  Ped=log( (pb$PD1_Ped-pb$Untreated_Ped)/(pb$Untreated_Ped+0.01) + 1 ),
                  row.names = rownames(pb))
  colnames(fc)<- paste0(colnames(fc), "_",program )
  return(fc)
})
fc<- do.call("cbind", all_fc)

## Stats- log TPM pseudobulked by sample + program, t.test
  ## Pseudobulk by sample + program
  seurat_use$sample<- ifelse(seurat_use$Age_Group=="Ped", seurat_use$sample, seurat_use$old.ident)
  seurat_use$pb<- paste0(seurat_use$sample, "__", seurat_use$Program_projected)
  pb<- pseudobulk_byVariable(seurat_use, seurat_use@meta.data, "pb")

  ## reformat- transpose + add program column
  pb_t<- as.data.frame(t(pb[GOI,]))
  pb_t$Program<- sapply(strsplit(rownames(pb_t), split="__"), function(x){x[2]})
  
  ## Add on treatment + age
  pb_t$Sample<- sapply(strsplit(rownames(pb_t), split="__"), function(x){x[1]})
  pb_t$Treatment<- as.character(plyr::mapvalues(pb_t$Sample, seurat_use$sample, seurat_use$Treatment, warn_missing = FALSE))
  pb_t$Age<- as.character(plyr::mapvalues(pb_t$Sample, seurat_use$sample, seurat_use$Age_Group, warn_missing = FALSE))
  
  ## Calculate mean TPM by age for untreated only
  mean_tpm_untreated<- pb_t %>% filter(Treatment=="Untreated") %>%group_by(Age, Program) %>% summarise(across(GOI, mean)) %>% as.data.frame()
  colnames(mean_tpm_untreated)<-c("Age", "Program", paste0("Untreated_", colnames(mean_tpm_untreated[3:ncol(mean_tpm_untreated)])))
  
  ## Merge untreated (mean by age/program) with treated (separate by sample)
  pb_pd1<- pb_t[pb_t$Treatment=="PD1",]
  pb_pd1$UniqueID<- paste0(pb_pd1$Age, "_", pb_pd1$Program)
  mean_tpm_untreated$UniqueID<- paste0(mean_tpm_untreated$Age, "_", mean_tpm_untreated$Program)
  pb_merge<- merge(pb_pd1[,!colnames(pb_pd1) %in% c("Age", "Program", "Treatment")],
                   mean_tpm_untreated[!colnames(mean_tpm_untreated) %in% c("Age", "Program", "Treatment")],
                   by="UniqueID")
  
  ## Calculate fold change- TPM treated - mean for untreated from that age.program / mean for untreated from that age.program
  fc_bySample<- sapply(GOI, function(x){
    tmp<- pb_merge[,c(x, paste0("Untreated_", x))]
    fc<-log((tmp[[x]] - tmp[[paste0("Untreated_", x)]]) / (tmp[[paste0("Untreated_", x)]]+0.01) + 1 )
    return(fc)
  })
  fc_bySample<- as.data.frame(fc_bySample)
  fc_bySample$UniqueID<- pb_merge$UniqueID
  fc_bySample$Age<-  sapply(strsplit(fc_bySample$UniqueID, split="_"), function(x){x[1]})
  fc_bySample$Program<-  sapply(strsplit(fc_bySample$UniqueID, split="_"), function(x){x[2]})
  
  ## ttest comparing fold changes
  pvalues_all<- lapply(unique(fc_bySample$Program), function(program){
    tmp<- fc_bySample[fc_bySample$Program==program,]
    pvalues<-sapply(GOI, function(gene){
      t.test(tmp[tmp$Age=="Ped", gene], tmp[tmp$Age=="Adult", gene])$p.value
    })
    pvalues<- as.data.frame(pvalues)
    colnames(pvalues)<- program
    return(pvalues)
  })
  pvalues_all<- do.call("cbind", unname(pvalues_all))
  padjust_all<- as.data.frame(apply(pvalues_all, 1, function(x){p.adjust(x)}))
  
  ## reformat for adding to plot
  padjust_all$Program<- rownames(padjust_all)
  padjust_plot<- melt(padjust_all)
  colnames(padjust_plot)<- c("Program", "Gene", "padj")
  
  ## For each group (program, gene, ped/adult- will always be both), set whether p is <0.05
  padjust_plot<- rbind(padjust_plot, padjust_plot)
  padjust_plot$XAxis<- c(rep("Adult", nrow(padjust_plot)/2), rep("Ped", nrow(padjust_plot)/2))
  padjust_plot$Box<- ifelse(padjust_plot$padj<0.1, TRUE, FALSE)
  padjust_plot$FacetWrap<- padjust_plot$Program
  
  write.csv(padjust_all, file=paste0())

## Plot
colnames(fc)<- gsub("\\.", "\n", colnames(fc))
myHeatmap(fc, GOI=GOI, min.value = -2, max.value = 2, facetWrap = TRUE, addBoxes = padjust_plot) +
  scale_fill_gradientn(colors=c( "khaki3", "lemonchiffon1","white","palevioletred1", "palevioletred2", "violetred4"))
ggsave(paste0(figure_dir, "Heatmap_foldChange_GOI_CD8_byProgram.pdf"),device="pdf", width=12, height=4)

## RAW DATA EXPORT
write.csv(fc, file="../..//notUploaded/ImmuneDraft_Rebecca/raw_results_files/fold.change_pd1.itn_adult.ped.csv")
```



## Pseudobulked heatmap of expression of GOI (Myeloid)
```{r}
GOI<- c("PVR", "PVRL2","CD274", "CD80", "CD86", "IFNG")
sample_center_method="bySample"

## By sample
all_plots<- lapply(c("myeloid"), function(x){
  tmp<- seurat_list[[x]]
  tmp$Treatment<- gsub("Treated", "PD1", tmp$Treatment)
  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$Treatment, "_", tmp$Age_Group)
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  ## center for each sample separately
  if(sample_center_method=="bySample"){
      center_bySample<- lapply(unique(tmp$Treatment), function(sample){
    pb_tmp<- pb[,grepl(sample, colnames(pb))]
    if(class(pb_tmp)=="numeric"){
      pb_tmp<- as.data.frame(pb_tmp)
      colnames(pb_tmp)<- colnames(pb)[grepl(sample, colnames(pb))]
    }
    pb_list_tmp<- NormCenter(pb_tmp)
    pb_center_tmp<- pb_list_tmp$center_data
    return(pb_center_tmp)
  })
  pb_center<- do.call("cbind", center_bySample)
  } else if(sample_center_method=="byAll"){pb_center<- NormCenter(pb)$center_data}
  
  
  ## Plot
  myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE,
            orderFactors = c("Ped", "Adult"), orderSubtypes = c("Untreated", "PD1")) +
    scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=1)
ggsave(paste0(figure_dir, "Heatmap_GOI_Myeloid.png"), width=7, height=5)

```

# Heatmap of DEGs between treated/untreated in ped- same pattern in adult?
```{r}
pvalue_thresh<- 0.1
shared_degs<- readRDS(paste0("../02_Merge_SS2.tenX/analysis/SharedDEGs/shared_degs_tenx.ss2_pvalueThresh", pvalue_thresh, ".Rds"))
shared_degs<- lapply(shared_degs, function(x){x[x$sameLogFCDirection=="Same",]})

sample_center_method="byAll"
all_plots<- lapply(names(seurat_list), function(x){
  tmp<- seurat_list[[x]]
  markers_use<-shared_degs[[x]]
  
  ## subset to adult
  tmp<- subset(tmp, Age_Group=="Adult")
  
  ## subset to programs with degs
  tmp<- subset(tmp, Program_projected %in% unique(markers_use$program))
  markers_use<- markers_use[markers_use$gene %in% rownames(tmp),]
  markers_use<- markers_use[order(markers_use$program, markers_use$SS2_avg_log2FC),]

  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$Treatment, "_", gsub("\\.", "\n", gsub("_", "\n", tmp$Program_projected)))
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  
   ## center for each sample separately
  if(sample_center_method=="bySample"){
      center_bySample<- lapply(unique(tmp$Treatment), function(sample){
    pb_tmp<- pb[,grepl(sample, colnames(pb))]
    if(class(pb_tmp)=="numeric"){
      pb_tmp<- as.data.frame(pb_tmp)
      colnames(pb_tmp)<- colnames(pb)[grepl(sample, colnames(pb))]
    }
    pb_list_tmp<- NormCenter(pb_tmp)
    pb_center_tmp<- pb_list_tmp$center_data
    return(pb_center_tmp)
  })
  pb_center<- do.call("cbind", center_bySample)
  } else if(sample_center_method=="byAll"){pb_center<- NormCenter(pb)$center_data}
  
  
  
  ## Plot
  myHeatmap(pb_center, GOI=unique(markers_use$gene), min.value = -3, max.value = 3, 
                orderFactors =  gsub("\\.", "\n",gsub("_", "\n", unique(markers_use$program))),
            orderSubtypes =c("Untreated", "PD1")) +
    scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=3)
ggsave(paste0(figure_dir, "Heatmap_SharedDEGs_pvalueThresh",pvalue_thresh, ".pdf"), width=17, height=5)
```


## Pseudobulked heatmap of expression of GOI, by programs 
```{r}
GOI<- c("KLRB1", "SELPLG", "PTGER2", "PDCD1", "TIGIT", "CTLA4", "ITGAE", "GZMK", "GNLY", "NKG7", "GZMB", "GZMA", "GADD45B")

## T cells
  ## By treatment
  all_plots<- lapply(c("cd4", "cd8"), function(x){
    tmp<- seurat_list[[x]]
    
    ## Create pseudobulking variable
    tmp$pb<- paste0(tmp$Treatment, "_", gsub("\\.", "\n", gsub("_", "\n", tmp$MaxScoringProgram)))
    pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
    
    pb_center<- NormCenter(pb)$center_data
    
    ## Plot
    myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE, orderSubtypes = c("Untreated", "Treated")) +
      scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
      ggtitle(x)
  })
  cowplot::plot_grid(plotlist=all_plots, ncol=2)
  ggsave(paste0(figure_dir, "Heatmap_GOI_byProgram_Tcells.png"), width=28, height=7)
  
    ## By treatment + primary/recurrent
    all_plots<- lapply(c("cd4", "cd8"), function(x){
      tmp<- seurat_list[[x]]
      
      ## Create pseudobulking variable
      tmp$pb<- paste0(tmp$condition, "_", gsub("\\.", "\n", gsub("_", "\n", tmp$MaxScoringProgram)))
      pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
      
      pb_center<- NormCenter(pb)$center_data
      
      ## Plot
      myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE, orderSubtypes = c("GBM.new", "GBM.rec", "GBM.PD1")) +
        scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
        ggtitle(x)
    })
    cowplot::plot_grid(plotlist=all_plots, ncol=2)
    ggsave(paste0(figure_dir, "Heatmap_GOI_byProgram_Primary.Recurrent_Tcells.png"), width=28, height=7)


## Myeloid
  GOI<- c("HAMP", "LYVE1", "NINJ1", "RHOB", "SLC2A5", "STAB1", "IL32")
  
  tmp<- seurat_list$Myeloid
  
  ## Create pseudobulking variable
  tmp$pb<- paste0(tmp$Treatment, "_", gsub("\\.", "\n", gsub("_", "\n", tmp$MaxScoringProgram)))
  pb<- pseudobulk_byVariable(tmp, tmp@meta.data, "pb")
  
  pb_center<- NormCenter(pb)$center_data
  
  ## Plot
  myHeatmap(pb_center, GOI=GOI, min.value = -3, max.value = 3, facetWrap = TRUE, orderSubtypes = c("Untreated", "Treated")) +
    scale_fill_gradient2(high="red", low="blue", mid="white", name="Centered log2(TPM)")+
    ggtitle("Myeloid")
  
  ggsave(paste0(figure_dir, "Heatmap_GOI_byProgram_Myeloid.png"), width=14, height=5)
```

## NOT IN USE
## Based on heatmap- boxplot of mean gene expression for GOI within program of interest
```{r}
tmp<- seurat_list$CD4


## Boxplot
goi_plot<- c("PDCD1", "GNLY", "TIGIT", "CTLA4", "GZMB", "SELPLG", "PTGER2", "KLRB1", "GADD45B")
all_plots<-lapply(goi_plot, function(gene){
  tmp$GOI<- as.data.frame(tmp@assays$RNA@data[gene,])
  mean_goi<- tmp@meta.data %>% group_by(sample, MaxScoringProgram) %>% summarise(mean_expr=mean(GOI)) %>% as.data.frame()
  mean_goi$Treatment<- plyr::mapvalues(mean_goi$sample, tmp$sample, tmp$Treatment, warn_missing = FALSE)
  mean_goi$Treatment<- factor(mean_goi$Treatment, levels=c("Untreated", "Treated"))
  
  ggplot(mean_goi, aes(x=Treatment, y=mean_expr, fill=Treatment))+
    geom_boxplot()+
    geom_point()+
    ylim(0,max(mean_goi$mean_expr))+
    facet_grid(cols=vars(MaxScoringProgram))+
    stat_compare_means(method="t.test")+
    theme_bw()+
    scale_fill_manual(values=c(Untreated="purple", Treated="orange"))+
    theme(axis.text.x = element_text(angle=45, hjust=1),
          axis.text = element_text(face="bold", color="black"),
          axis.title = element_text(face="bold", color="black"),
          strip.text = element_text(face="bold", color="black", size=10),
          legend.position = "none")+
    xlab("")+
    ylab(paste0("Mean log2 TPM: ", gene))
})
cowplot::plot_grid(plotlist=all_plots, ncol=1)
ggsave(paste0(figure_dir, "Boxplot_AllGOI_MeanExprBySample_PD1vUntreated_CD4.png"), width=10, height=4*length(goi_plot))

## VlnPlot
VlnPlot(tmp, group.by = "MaxScoringProgram",stack=TRUE , flip = TRUE,
        features=c("PDCD1", "TIGIT", "CTLA4", "ITGAE", "GZMK", "GNLY", "NKG7", "GZMA", "SELPLG", "PTGER2","KLRB1", "GADD45B"), 
        combine=TRUE, fill.by="ident", split.by = "Treatment", split.plot = TRUE)+
  scale_fill_manual(values= c(Untreated="purple", Treated="orange"))
ggsave(paste0(figure_dir, "Vln_GOI_byTreatment_CD4.png"), width=7, height=8)
```

## Score for cytoxic/dysfunctional genesets, plot by treatment
```{r}
## Add nature 2020 scores (dysfunctional, cyto, dys, memory)
all_seurat<- lapply(all_seurat, function(x){
  x<-AddModuleScore(x, nature_2020_genes, name=paste0(names(nature_2020_genes), "_score"))
  colnames(x@meta.data)<-ifelse(grepl(paste(paste0(names(nature_2020_genes), "_score"), collapse = "|"), colnames(x@meta.data)),
                                gsub("[0-9]", "", colnames(x@meta.data)),
                                colnames(x@meta.data))
  return(x)
})

## Plot
all_plots<- lapply(names(all_seurat), function(x){
  seurat<- all_seurat[[x]]
  seurat$sample<- factor(seurat$sample, sample_order[sample_order %in% unique(seurat$sample)])
  DotPlot(seurat, group.by = "sample", features=paste0(names(nature_2020_genes), "_score"))+
    scale_color_gradient2(low="blue", mid="grey", high="red")+
    coord_flip()+
    theme(axis.text.x = element_text(angle=45, hjust=1))+
    ggtitle(x)
})
cowplot::plot_grid(plotlist=all_plots, ncol=1)
ggsave(paste0(figure_dir, "DotPlot_TcellStateScores_CD4.CD8.png"), width=13, height=8)
```


## RAW DATA EXPORT
```{r}
## Pediatric SS2- for merging with "ped_cellwise_annotations.csv"
ped_cellwise_annotations<- read.csv("../../notUploaded/ImmuneDraft_Rebecca/raw_results_files/ped_cellwise_annotations.csv", row.names = "X")

ped_ss2<- lapply(seurat_list, function(x){subset(x, Age_Group=="Ped" & Treatment=="PD1" & Sequencing=="SS2")})
ped_ss2_tmp<- merge(ped_ss2$cd4, ped_ss2$cd8)
ped_ss2<- merge(ped_ss2_tmp, ped_ss2$myeloid)
  
  ## Add sort info
  ped_ss2$sort<- ifelse(grepl("CD3", ped_ss2$orig.ident), "CD3", 
                        ifelse(grepl("CD45", ped_ss2$orig.ident), "CD45",
                        ifelse(grepl("BT2080", ped_ss2$orig.ident), "CD3", "issue")))
  ped_ss2$subtype<- "GBM"
  
  ## Add mg/mo score/annot- just null
  ped_ss2$Microglia_Macrophage_annot<- "NA"; ped_ss2$Microglia_Macrophage_Score<- "NA"
  
  ## add clonal- just null. never used (too few samples)
  ped_ss2$tcr_reconstructed<- "NA"; ped_ss2$clonal_group<- "NA"; ped_ss2$group_size<-"NA"; ped_ss2$clonal<- "NA"
  
  ped_ss2_df<- ped_ss2@meta.data[,c("sample", "sort", "subtype", "broad_annot_projected","detailed_annot_projected_manual",
                                    "Program_projected", "Microglia_Macrophage_annot","Microglia_Macrophage_Score",
                                    "tcr_reconstructed", "clonal_group", "group_size", "clonal")]
  ped_ss2_df<- ped_ss2_df %>% dplyr::rename(broad_annot=broad_annot_projected, detailed_annot=detailed_annot_projected_manual,
                                            Program=Program_projected) %>% as.data.frame()
  
  ## Add treatment info- to PD1 (to be added here) and full SS2 cohort
  ped_ss2_df$Treatment<- "AntiPD1"
  ped_cellwise_annotations$Treatment<- "ITN"
  
  ## Combine
  ped_cellwise_annotations<- rbind(ped_cellwise_annotations, ped_ss2_df)
  
  write.csv(ped_cellwise_annotations, file="../../notUploaded/ImmuneDraft_Rebecca/raw_results_files/ped_cellwise_annotations.csv")
  



## Pediatric 10X- for merging with "ped_cellwise_annotations.csv"
ped_10XGenomics_cellwise_annotations<- read.csv("../../notUploaded/ImmuneDraft_Rebecca/raw_results_files/ped_10XGenomics_cellwise_annotations.csv", row.names = "X")

ped_tenx<- lapply(seurat_list, function(x){subset(x, Age_Group=="Ped" & Treatment=="PD1" & Sequencing=="TenX")})
ped_tenx_tmp<- merge(ped_tenx$cd4, ped_tenx$cd8)
ped_tenx<- merge(ped_tenx_tmp, ped_tenx$myeloid)
  
  ## Update/clarify any metadata columns
  ped_tenx$broad_annot_projected_manual<- ifelse(is.na(ped_tenx$broad_annot_projected_manual), "Tcell", ped_tenx$broad_annot_projected_manual)
  
  
  ## add clonal
  ped_tenx$tcr_reconstructed<- ifelse(ped_tenx$v_gene != "NA", "Yes", "No")
  ped_tenx$clonal<- ifelse(ped_tenx$tcr_reconstructed=="No", "NA", ifelse(ped_tenx$clonotype_size>1, "clonal", "nonclonal"))
  
  ped_tenx_df<- ped_tenx@meta.data[,c("sample",   "broad_annot_projected_manual","detailed_annot_projected_manual",
                                    "Program_projected", 
                                    "tcr_reconstructed", "v_gene", "j_gene", "raw_clonotype_id", "clonotype_size", "clonal")]
  ped_tenx_df<- ped_tenx_df %>% dplyr::rename(broad_annot=broad_annot_projected_manual, detailed_annot=detailed_annot_projected_manual,
                                            Program=Program_projected) %>% as.data.frame()
  
  ## Add treatment info- to PD1 (to be added here) and full tenx cohort
  ped_tenx_df$Treatment<- "AntiPD1"
  ped_10XGenomics_cellwise_annotations$Treatment<- "ITN"
  
  ## Combine
  ped_10XGenomics_cellwise_annotations<- rbind(ped_10XGenomics_cellwise_annotations, ped_tenx_df)
  
  write.csv(ped_10XGenomics_cellwise_annotations, file="../../notUploaded/ImmuneDraft_Rebecca/raw_results_files/ped_10XGenomics_cellwise_annotations.csv")

## Adult- generate new file
adult<- lapply(seurat_list, function(x){subset(x, Age_Group=="Adult")})  
adult_tmp<- merge(adult$cd4, adult$cd8)
adult<- merge(adult_tmp, adult$myeloid)

adult_df<- adult@meta.data[,c("sample", "broad_annot", "detailed_annot", "Program_projected")]
adult_df<- adult_df %>% dplyr::rename(Program=Program_projected) %>% as.data.frame()
write.csv(adult_df, file="../../notUploaded/ImmuneDraft_Rebecca/raw_results_files/adultPD1_cellwise_annotations.csv")
```

